Config File Simplification (#9)
eslint/rfcs: designs/2019-config-simplification/README.md

The only reason that these requests are difficult to implement in ESLint is because of how complex the `.eslintrc` configuration format is. Any changes made to any part of `.eslintrc` processing end up affecting millions of ESLint installations, so we have ended up stuck. The complicated parts of `.eslintrc` include:

1. Resolution behavior of modules (`plugins`, `parser`, `extends`)
1. Merging of cascading config files in a directory structure
1. Merging of config files via `extends`
1. Overriding of configuration via `overrides`

All of these complexities arise from `.eslintrc` format of describing what should happen rather than how it should happen. We keep trying to anticipate how users want things to happen and guessing wrong has led to increased complexity.

This RFC proposes a fresh start for configuration in ESLint that takes into account all of the requests we've received over the years. By simplifying and stripping configuration we can actually make a configuration system that is flexible enough to accomodate changes in the future and easy enough to implement that we won't be afraid to do so.

## Detailed Design

Design Summary:

> 1. Introduce a new `.eslint.js` configuration file format

Reviewer: One thing to consider - a lot of popular JS tooling projects have started using the `project-name.config.js`  filename format (including [Webpack](https://webpack.js.org/configuration/), [Babel](https://babeljs.io/docs/en/configuration#babelconfigjs), and [Prettier](https://prettier.io/docs/en/configuration.html)). Glancing at `.eslint.js` quickly, I think it's very easy to misread it as `.eslintrc.js` if that's what you're used to. `eslint.config.js` would follow what is starting to become a convention and could make it clearer that it's a new file format.

Author: Oh, I wasn't aware of that. I'm happy to change the proposal to `.eslint.config.js`. Great feedback.

Reviewer 2: > I'm happy to change the proposal to `.eslint.config.js`.

Just to clarify, the suggestion from https://github.com/eslint/rfcs/pull/9#discussion_r249959649 and https://github.com/eslint/rfcs/pull/9#discussion_r249968780 was to call the config `eslint.config.js` rather than `.eslint.config.js` (i.e. no leading dot). I'm happy to discuss other options but I'm not sure whether the leading dot here was a typo or an alternative suggestion.

Reviewer 3: Matches what cosmiconfig looks for as well https://github.com/davidtheclark/cosmiconfig/blob/master/README.md

Author: Oops, I misread that. I'll change to `eslint.config.js`. Thanks for catching that.

1. Searching for `.eslint.js` starts from the current working directory and continues up
1. All `.eslint.js` files are treated as if they have `root: true`
1. There is no automatic merging of config files
1. The `.eslint.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
1. Remove the concept of environments (`env`)
1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
1. Remove `--rulesdir`
1. Create a `@eslint/config` package to help users merge configs

### The `.eslint.js` File

> The `.eslint.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 

Reviewer 2: Minor: Is there a reason that a config object is assigned to `exports.config` rather than `module.exports`?

Author: Two reason:

1. I'm [not a fan](https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/) of default exports.
1. I'd like to leave open the possibility of exporting other things from a config file without forcing it into the primary config structure.

Reviewer 2: In that case, should the examples that depend on shareable configs be updated to use `require('eslint-config-airbnb').config` to locate a shareable config, rather than `require('eslint-config-airbnb')`? Or is the intent that end-user configs use `module.exports.config` while shareable configs use `module.exports`?

Author: Because we are just passing objects in, it doesn't matter. That's one of the downsides of this proposal: shareable configs and plugins basically have no enforced structure anymore. It will be up to each shareable config to define how it should be extended.

I'd still like to encourage shareable configs to use `module.exports` for simplicity, but there would be no restrictions around that.

Author: Adding a FAQ about this.

```js
exports.config = {
    files: "*.js",

>     ignore: "*.test.js",

Reviewer 4: `files` and `ignore` should be `Array` right? (based on current functionality)

Author: I was imagining these being strings that used `;` as a separator. I don't know that having them be arrays gives much of an advantage, but let's see what others think.

Reviewer 5: Letting it be an array means that it can be modified with general JavaScript array methods (like filtering based on environment, concatenation based on multiple child configs, etc.) Also easier to inspect or validate, and will be better formatted by autoformatters like Prettier.

Author: These have been changed to arrays per the feedback.

    globals: {},
    settings: {},
    processor: object,
    parser: object,
    parserOptions: {},
    ruledefs: {},
    rules: {}
};
```

The following keys are new to the `.eslint.js` format:

> * `files` - **Required.** Determines the global pattern that this configuration applies to and also is used as the default pattern to search for when ESLint is executed on the command line.

Reviewer 2: To clarify, does this mean that when running `eslint` without any arguments, all the files from the config are used (rather than displaying the help text)? This seems like the implication but I didn't see it mentioned in the "File Pattern Resolution" section.

Author: That was my initial inclination, but I had second thoughts and forgot to remove it from here.

> * `ignore` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this glob pattern are subtracted from the files specified in `files`.

Reviewer 2: Would `files` and `ignore` be inherited by 'require('@eslint/config').merge`? It seems potentially confusing to have a shareable config change what files are being linted, although maybe it does have use cases.

Author: Yes. It does, see @mysticatea's [comment](https://github.com/eslint/rfcs/pull/9#discussion_r250037797) .

* `ruledefs` - Contains definitions for rules grouped by a specific name. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.

The following keys are specified the same as in `.eslintrc` files:

* `globals`
* `settings`
* `rules`
* `parserOptions`

The following keys are specified differently than in `.eslintrc` files:

* `parser` - an object in `.eslint.js` files (a string in `.eslintrc`)
* `processor` - an object in `.eslint.js` files (a string in `.eslintrc`)

Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `.eslint.js`, these are all objects, requiring users to manually specify the objects to use.

The following keys are invalid in `.eslint.js`:

* `env` - responsibility of the user

> * `extends` - responsibility of the user

Reviewer 6: I think removing this, significantly complicates what I think is the most common scenario: I want to use AirBnB config, but I don't like 2-3 rules in it, and want to change their values. Before - I could just extend AirBnB and specify new config for 3 rules I didn't like. Now I need to require AirBnB config, retrieve rules object and override in place 3 rules I don't like. While it's not a huge difference, it does make it a bit more complicated, and if my assumption is correct (that it's the most common scenario) it will add a huge number of combined hours to creating basic configuration when adding up the time across all of the users.

Reviewer 6: I see that this is addressed by `@eslint/config` package. Not sure that's simple enough to explain it to somebody without showing them the code.

Author: Yeah, the `@eslint/config` package is intended to solve this problem. 

For new projects, I envision using `--init` to setup a basic config that uses `Config.create()` by default to setup a file that is easier to modify.

Author: `extends` has been added back into this proposal.

> * `overrides` - responsibility of the user

Reviewer 7: so if a user wants to lint types of files, let's say:
+ lint `fileA` using `configA`;
+ lint `fileB` with `configB`;

he/she has to create 2 configs?(atm,  we are using overrides: https://github.com/eslint/eslint/blob/master/.eslintrc.js#L29)

an alternate way is allowing the `.eslint.js` to be `array|object`:
```js
module.exports = [
{ files: ["fileA"], ...},
{ files: ["fileB"], ...}
];
```

* `plugins` - replaced by `ruledefs`
* `root` - always considered `true`

Each of these keys represent different ways of augmenting how configuration is calculated and all of that responsibility now falls on the user.

#### Referencing Plugin Rules

The `plugins` key in `.eslintrc` was an array of strings indicating the plugins to load, allowing you to specify processors, rules, etc., by referencing the name of the plugin. It's no longer necessary to indicate the plugins to load because that is done directly in the `.eslint.js` file. For example, consider this `.eslintrc` file:

```yaml
plugins:
  - react

rules:
  react/jsx-uses-react: error
```

This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.

In `.eslint.js`, the same configuration is achieved using a `ruledefs` key:

```js
const reactPlugin = require("eslint-plugin-react");

exports.config = {
    ruledefs: {

>         react: reactPlugin.rules

Reviewer 6: One thing to consider here is how the users will be able to search the internet for specific error they receive if they use shareable config and config author decided to rename the plugin. I.e.:

```js
const reactPlugin = require("eslint-plugin-react");

exports.config = {
  ruledefs: {
    "not-angular": reactPlugin.rules
  },
  rules: {
    "not-angular/jsx-uses-react": "error"
  }
}
```
As far as I understand, with configuration like that, error on the console will be reported as `not-angular/jsx-uses-react` which will make it hard to find which plugin error came from. Also, for overriding rules, you have to know what the author of shareable config decided to name plugins as well, which means reading shareable config source code.

Author: Yes, there is no more canonical name for a rule in this proposal. I'm not entirely sure that matters, though. I anticipate a lot of users would just use the plugin name as their namespace. We could encourage plugin authors to export a `namespace` property so there is a canonical name to refer to:

```js
const reactPlugin = require("eslint-plugin-react");

exports.config = {
  ruledefs: {
    [reactPlugin.namespace]: reactPlugin.rules
  },
  rules: {
    "not-angular/jsx-uses-react": "error"
  }
}
```

Would that make sense?

Reviewer 8: could maybe we just pass in `reactPlugin`, only, and let the schema of that object be a contract maintained between plugins and eslint (ie, not involving end users)?

Author: @ilyavolodin I think that if we help them with tools like `--init` then this becomes clearer. Eventually everyone will understand computed properties, so I'm not too worried about that.

@ljharb what type of contract are you envisioning? In this proposal, the concept for a formal plugin really goes away as you can pass any object that contains rules into `ruledefs`, so requiring plugins to look a certain way could restrict what otherwise is a wide-open playing field.

Reviewer 8: You could start with "an object with a `rules` property", but then that allows for future backwards-compatible expansion, without requiring any config changes for users (only core changes, and then corresponding plugin changes).

Author: Please see the FAQ I added about this.

Author: I'm adding a new section explaining how we can retain canonical rule names without making changes to this design.

    },
    rules: {

>         "react/jsx-uses-react": "error"

Reviewer 6: If we are making a departure from the current config, wouldn't it be easier to change the format to a nested object:
```js
rules: {
  react: {
    "jsx-uses-react": "error"
  }
}
```
This makes is a lot more machine readable, plus would be easier to override specific plugin configuration.

Author: :-1: I don't like that for a couple of reasons:

1. Makes it harder to reuse existing shareable configs because you know have to change the format.
2. With core rules that don't have a namespace, would we end up with core rule names at the same level as the namespace? I'm not a fan of mixing the meaning of key locations.

    }
};
```

Here, it is the `ruledefs` that assigns the name `react` to the rules from `eslint-plugin-react`. The reference to `react/` in a rule will always look up that value in the `ruledefs` key.

#### Referencing Parsers and Processors

In `.eslintrc`, the `parser` and `processor` keys required strings to be specified, such as:

```yaml
plugins: ["markdown"]
parser: "babel-eslint"
processor: "markdown/markdown"
```

In `.eslint.js`, you would need to pass the references directly, such as:

```js
exports.config = {
    parser: require("babel-eslint"),
    processor: require("eslint-plugin-markdown).processors.markdown
};
```

> In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never of question of where the modules will be resolved from.

Reviewer 6: ```suggestion
In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never a question of where the modules will be resolved from.
```

#### Applying an Environment

Unlike with `.eslintrc` files, there is no `env` key in `.eslint.js`. Users can mimmick the behavior of `env` by assigning directly to the `globals` key:

```js
const globals = require("globals");

exports.config = {
    globals: {
        MyGlobal: true,
        ...globals.browser
    }
};
```

This effectively duplicates the use of `env: { browser: true }` in ESLint.

**Note:** This would allow us to stop environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.

#### Extending Another Config

With the removal of `extends`, it's now up to the users to inherit from other configs. That can easily be done in the following manner:

```js
const standardConfig = require("eslint-config-standard");

> standardConfig.rules.semi = ["error", "always"];

Author: That's a good point, and actually what `Config.create()` was meant to address. I'll update that.

> exports.config = standardConfig;

Reviewer 2: Would the user still need to add a `files` key to the shareable config? Presumably the files being linted are project-specific.

Author: Good point! Shareable configs shouldn't be providing `files` or `ignores`. I'll add mention of that.

Author: Actually, upon reading this section over again, I can see a use case for shareable configs providing `files` and `ignore`. Shareable configs can be general, like the Airbnb style guide, or company-specific, in which case they might want to have these fields standardized.

Reviewer 9: Not sure if the extra complexity is worth it, but-- could/should we add either a config key, a CLI/CLIEngine option, or an argument to `Config.merge` etc., that will opt-in or opt-out of receiving globs from extended configs during config merge?

Author: My thinking was that people who didn't want to import `files` and `ignore` would just specify their own patterns that would overwrite the ones they were extending. We could certainly add a utility to help pick things out, but I think the necessity is based on how we decide to merge those fields. If we always overwrite, then I don't think we need a utility for this purpose.

Also, see @mysticatea's [comment below](https://github.com/eslint/rfcs/pull/9#discussion_r250037797) about how he uses file patterns in shareable configs currently.

#### Overriding Configuration Based on File Patterns

Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `.eslint.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:

```yaml
plugins: ["react"]
rules:
   react/jsx-uses-react: error
    semi: error

overrides:
    - files: "*.md"
      plugins: ["markdown"],
      processor: "markdown/markdown"
```

This can be written in `.eslint.js` as an array of two configs:

```js
exports.config = [
    {
        files: "*.js",
        ruledefs: {
            react: require("eslint-plugin-react").rules,
        },
        rules: {
            "react/jsx-uses-react": "error",
            semi: "error"
        }
    },
    {
        files: "*.md",
        processor: require("eslint-plugin-markdown").processors.markdown
    }
];
```

When ESLint uses this config, it will check each `files` pattern to determine which config applies. In this way, returning an array acts exactly the same as the array in `overrides`, with the exception that no configuration options are inherited.

#### Replacing `.eslintignore`

> Because there is only one `.eslint.js` file to consider, ESLint doesn't have to first search directories to determine its location. That allows `.eslint.js` to specify files to ignore directly instead of relying on `.eslintignore`. For backwards compatibility, users could create a config like this:

Reviewer 6: I'm not sure that's a correct assumption. Just because each config file basically acts as if `root: true` is specified, doesn't mean that users will choose not to create a separate config file in their `tests` directory, for example. They might even inherit manually from the higher config file by requiring it.

Author: That's true, though I would expect they would use a `files` pattern for that purpose rather than creating a separate file. I think my general use case assumption still makes sense here.

```js
const fs = require("fs");

exports.config = {
    files: "*.js",

>     ignore: fs.readFileSync(".eslintignore").replace(/\n/g, ";")

Reviewer 6: It would be nicer if ignore property would accept an array of strings, instead of a semi-column separated string.

Reviewer 10: I prefer:

```suggestion
    ignore: fs.readFileSync(".eslintignore").split("\n")
```

Reviewer 6: @nzakas I don't like artificial constructs. JavaScript has a very nice data structure for a list of strings, why invent a new way of recording them? It also improves machine readability of the config as well as makes it easier to see information at a glance. I don't see any downsides of using an array, but I see downside in using a single string (user has to parse it manually, if they want to modify it in the code).

Author: I guess my perspective is that people write glob patterns on the command line all time as a single string, and I'm not sure why I'd want to force people into writing extra syntax if it's not necessary.

That said, it seems like there's general agreement from the team that arrays would be better, so I'll make that change.

};
```

### The `@eslint/config` Package

Because some of the operations ESLint currently do are quite complicated, this design includes a utility package called `@eslint/config` that users can use for common tasks. This package contains the following utilities:

1. `Config` - a class that both wraps a configuration object to provide helper methods for manipulating configs and a container for static methods to work with configs.
1. `RuleLoader` - a class that aids in loading rules from a directory.

The `@eslint/config` package is intended to add back functionality that would be removed from ESLint with this design.

**Note:** It is not required to use this package with `.eslint.js`; this is an optional tool that users may choose to use for convenience.

#### `Config` Class

The `Config` class has this form:

```js
class Config {

    // pass in any object literal to wrap in Config
    constructor(object) {}

    // convenience method for easily setting rule severity
    setRuleSeverity(ruleId, severity) {}

    // determine if an object is a Config
    static isConfig(object) {}

    // create a new config based on another config

>     static create(baseConfig, overrideConfig) {}

Reviewer 10: Can I use an array as the `baseConfig`?

Currently, we can use `overrides` property in shared configuration. I'm a heavy user of that. 

```yml
root: true
extends:
  - plugin:@mysticatea/es2018
```

The above config has override settings for each `*.vue`, `*.ts`, `*.tsx`, `test/**`, `scripts/**`, `rollup.config.js`, and `webpack.config.js`.

Since new configuration system replaces the `overrides` property to an array of configs, it will be much if the `baseConfig` doesn't accept an array.

Reviewer 10: ...how about using an array instead of `assign`?

```js
const { configs } = require("@mysticatea/eslint-plugin")

exports.config = [
    // corresponding `extends: "plugin:@mysticatea/es2018"`
    ...configs.es2018,

    // corresponding `overrides`
    {
        files: "*.js",
        rules: {
            semi: ["error", "always"]
        }
    }
]
```

Because the current `overrides` merges all matched entries, I'm imaging the array config in the new system is same.
In this case, we can use it as substitude of `extends` property. Also, the string entry for embedded presets will be useful.

```js
exports.config = [
    "eslint:recommended",
    {
        files: "*.js",
        rules: {
            semi: "error"
        }
    }
]
```

Reviewer 10: So the solution I'm imagining is:

```js
// Merge all matched entries.
const actualConfig = Config.create(...configs.filter(config => match(filePath, config)))
```

This is a very simple rule: the setting at the last entry is prior to others.
I don't think that it's difficult to understand, as different from the current `overrides` property and `extends` property.

Author: I see, thanks for explaining. Can you take a look at the latest proposal that adds back `extends` and see if that addresses your concerns?

Author: Sorry, I just read this over again and I think I'm not understanding correctly. I think what you're saying is:

In the current system, if a shareable config has `overrides` then  those `overrides` are merged into a single list when `extends` is used, and if two `files` patterns match, then those overrides are merged together? Is that correct?

And in this proposal, because there are no overrides, you lose that ability?

If I'm understanding those points correctly, then I agree, that's a problem we need to solve. I'm not quite sure what you're referring to in that code example, though. Could you write out a more complete example?

Reviewer 10: > Can you take a look at the latest proposal that adds back extends and see if that addresses your concerns?

Sure. I guess that the `extends` property is not going to work properly. How does it work on the following case?

```js
exports.config = {
    extends: [
        // A shareable config which has overriding (= an array).
        [
            { files: "*.js", rules: { semi: "error" } },
            { files: "*.ts", parser: require("@typescript-eslint/parser")}
        ]
    ],
    files: "*.js"
    rules: { semi: ["error", "never"] }
}
```

Where does the extended TS config go?
How do I overwrite the TS config?

In the new system, every config entry has `files` property. I don't think that `extends` property with `overrids` functionality makes sense.

----

> In the current system, if a shareable config has overrides then those overrides are merged into a single list when extends is used, and if two files patterns match, then those overrides are merged together? Is that correct?
>
> And in this proposal, because there are no overrides, you lose that ability?

Yes, you are right.

----

> Could you write out a more complete example?

Sure. My code was the code to determine the actual config in ESLint core.

```js
/** Load eslint.config.js */
function loadConfig(cwd) {
    const configOrConfigs = require(path.resolve(cwd, "eslint.config"))
    const configs = Array.isArray(configOrConfigs) ? configOrConfigs : [configOrConfigs]
    // TBD: Validation
    return configs
}

/** Get actual config for a given file. */
function getActualConfig(configs, filePath) {
    return Config.assign({}, ...configs.filter(c => match(c, filePath)))
}
```

There is an assumption: both `extends` and `overrides` properties do not exist.

For each linting target file, ESLint merges all matched config entries in linear. This simple behavior satisfies all of `extends` functionality, `overrides` functionality, and overwriting for each project.

For example:

```js
exports.config = [
    // A shareable config which has overriding (= an array).
    ...[
        { files: "*.js", rules: { semi: "error" } },
        { files: "*.ts", parser: require("@typescript-eslint/parser")}
    ],
    // For this project.
    {
        files: "*.js"
        rules: { /* ... */ }
    },
    {
        files: "*.ts"
        rules: { /* ... */ }
    },
]
```

And the examples of https://github.com/eslint/rfcs/pull/9#discussion_r250039318 are valid.

Author: Ah ok, I see what you mean. Thanks for explaining. Let me think about this for a bit and I'll respond back here when I've fully thought this through.

Author: Okay, after getting some good sleep, I think I understand this fully now. Your suggestion makes a lot of sense and I'm working on getting it into this proposal.

    static assign(receiverConfig, ...supplierConfigs) {}
}
```

A new instance of `Config` is created like this:

```js
const { Config } = require("@eslint/config");

const config = new Config({
    files: "*.js",
    ruledefs: {
        react: require("eslint-plugin-react").rules,
    },
    rules: {
        "react/jsx-uses-react": "error",
        semi: ["error", "always"]
    }
});
```

After that point, you can easily change the severity of a rule like this:

```js
config.setRuleSeverity("semi", "warn");
```

In this way, users don't need to modify an array to change a rule's severity, which is a major downside of this design when using a vanilla object structure as a config.

The static methods on `Config` make it easy to create and modify configs. For example, suppose you want to create a config based on `eslint-config-standard` but would prefer `semi` be a warning instead of an error, you can do so using `Config.create()`:

```js
const standardConfig = require("eslint-config-standard");
const { Config } = require("@eslint/config");

const myConfig = Config.create(standardConfig, {
    rules: {
        semi: "warn"
    }
});

exports.config = myConfig;
```

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Author: Yup. It's solved the exact same way. The last one wins.

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Reviewer 2: I'm not sure I understand. What would be the result of merging two configs like this?

```js
// config 1
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("rule-1"),
      "my-rule2": require("rule-2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option1"]
    "foo/my-rule2": "error"
  }
};
```

```js
// config 2
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
  }
};
```

```js
Config.create(config1, config2) // => ?
```

Both configs are individually valid, but it's not clear how merging the configs would create another valid config. (Based on the "last one wins" rule, it seems like this would result in a config that (a) references a rule called `foo/my-rule2` even though no definition for such a rule is provided, and (b) provides the `"option1"` option to the rule at `require("some-other-rule1")`, which might not be a valid option for that rule.)

I'm not sure what you mean by "it's solved the exact same way" because I don't think we've solved this problem before, aside from the solution in #5.

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Author: "It's solved the same way" -> Even with `.eslintc`, it's the last config whose settings win. So in an `extends` array, the last config's settings override everything that came before it when there is a conflict, and then the config with `extends` takes priority over everything we calculate from `extends`.

`Config.create(a, b)` effectively creates `merge({ extends: a }, b)` (where `merge()` is our current merging strategy). So in your example, the resulting config would be:

```js
exports.config = {
    extends: config1,
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
  }
};
```

Or, if we hydrate the config to remove `extends`, the resulting config would look like this:

```js
// config 1
exports.config = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1")
      "my-rule2": require("rule-2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
    "foo/my-rule2": "error"
  }
};
```

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Reviewer 2: I see. What about this example?

<details>
<summary>Example</summary>

```js
// config 1
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("rule-1"),
      "my-rule2": require("rule-2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option1"]
    "foo/my-rule2": "error"
  }
};
```

```js
// config 2
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1"),
      "my-rule2": require("some-other-rule2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
  }
};
```

```js
module.exports.config = { extends: [config1, config2] }; // => What is the flattened equivalent?
```

</details>

The only difference between this example and the previous example is that config 2 now defines a `foo/my-rule2` rule, but does not configure it.

(Note: The existing merging strategy did not need to handle cases like this, because we required plugin names to be globally unique.)

I also notice that in the final merged config, the ruledef for `foo` contains two rules (`my-rule1` and `my-rule2`) from two different sources (one from each config). I'm concerned that this would create a confusing debugging experience for users; for example, if they need to report a bug to `eslint-plugin-foo`, it might not be obvious which version a particular rule is coming from. (Additionally, since exporting ruledefs under the name `foo` would no longer require someone to control the `eslint-plugin-foo` package on npm, it seems like having different plugins with the same name would become more common in practice.)

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Author: The existing merge strategy is that any top-level key that contains an object value has its properties merged, with the last one winning whenever those properties conflict. That's the same approach we'd use in this design for `ruledefs` (same merging strategy as `globals`, for example).

For this example, the hydrated result would be:

```js
// config 1
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1"),
      "my-rule2": require("some-other-rule2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
    "foo/my-rule2": "error"
  }
};
```

> Additionally, since exporting ruledefs under the name foo would no longer require someone to control the eslint-plugin-foo package on npm, it seems like having different plugins with the same name would become more common in practice.

As I've mentioned elsewhere, this is one of the downsides of this design: we lose control over package names and structures so we can no longer enforce anything. A lot of the issues you describe will have to be handled through documentation now. That's also why we need something like `ruledefs` so the end user can define the namespace for rules on their own.

I think we can mitigate some of this by encouraging people to add URLs to rule meta information and creating a new core formatter that outputs those URLs in some meaningful way.

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Reviewer 2: > For this example, the hydrated result would be:
> 
> ```js
> // config 1
> module.exports = {
>   ruledefs: {
>     foo: {
>       "my-rule1": require("some-other-rule1"),
>       "my-rule2": require("some-other-rule2")
>     }
>   },
>   rules: {
>     "foo/my-rule1": ["error", "option2"]
>     "foo/my-rule2": "error"
>   }
> };
> ```

Thanks for clarifying. This seems like a major problem to me. The config `"foo/my-rule2": "error"` was a configuration for the rule at `require("rule-2")`, but after configs get merged, that rule configuration ends up getting applied to a different rule at `require("some-other-rule2")`. If the two rules have incompatible schemas, this could result in a confusing config error when extending two different shareable configs.

In other words, I don't think this proposal currently offers a solution to the issues discussed in https://github.com/eslint/eslint/issues/3458 where two shareable configs depend on different versions of a plugin.

---

I see a few possible workarounds. In increasing order of complexity:

1. Raise a fatal error when merging two configs if they have overlapping `ruledefs` keys. Add a utility function to `@eslint/config` that renames a particular ruledef key in a config to something else, so that the end user can avoid these conflicts manually when merging configs. For example, an end-user's config might look like this:

    <details>
    <summary>Example config with solution 1</summary>

    ```js
    const { renameRuleDef } = require("@eslint/config");
    module.exports = {
        extends: [
            require("config1"),
            renameRuleDef(require("config2"), { from: "foo", to: "other-foo" })
        ],
        rules: {

            // configures the rule at require("rule-2")
            "foo/my-rule2": "error",

            // configures the rule at require("some-other-rule2")
            // (since that rule now appears in an "other-foo" ruledef)
            "other-foo/my-rule2": "error"
        }
    }
    ```
    </details>

    As a disadvantage, creating a config like this would probably be difficult for users who don't understand the details of how ruledefs work.

1. Introduce a rule-namespacing solution along the lines of https://github.com/eslint/rfcs/pull/5. We could probably avoid much of the complexity of #5 since this RFC removes so many features, but if we took this solution we might want to require users to name their `extends` clauses, so that users could refer to edges in the config tree. For example:

    <details>
    <summary>Example config with solution 2</summary>

    ```js
    module.exports = {
        // Note: `extends` is now an object, not an array
        extends: {
            config1: require("config1"),
            config2: require("config2")
        },
        rules: {

            // configures rules in the `foo` ruledef defined in `config1`
            "config1::foo/my-rule2": "error",

            // configures rules in the `foo` ruledef defined in `config2
            "config2::foo/my-rule2": "error",

            // If the "bar" ruledef is unambiguous, a `::` prefix is not required.
            "bar/my-rule3": "error"
        }
    }
    ```
    </details>

    As a disadvantage, this solution would require users to give explicit names to their extended configs (since `extends` would be an object) even in cases where no ambiguity was present. Note that this solution could not be added on after-the-fact without another big breaking change, since it would change the structure of `extends` clauses.

1. Combine the `ruledefs` and `rules` config options, so that users would load rules directly by symbols or object identity (similar to how users load parsers and processors in this proposal). This would avoid all naming conflicts because we wouldn't need to give rules user-declared names. For example:

    <details>
    <summary>Example config with solution 3</summary>

    ```js
    const fooPlugin = require("eslint-plugin-foo");
    module.exports = {
        rules: {
            [fooPlugin.rules["my-rule1"].KEY]: "error"
        }
    };
    ```
    </details>

    Solution 3 would probably introduce more complexity than it's worth, so I wouldn't recommend it at this point, but I'm mentioning it for the sake of completeness.

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Author: > The config "foo/my-rule2": "error" was a configuration for the rule at require("rule-2"), but after configs get merged, that rule configuration ends up getting applied to a different rule at require("some-other-rule2").

That's a good point. Would it make sense to throw an error if `foo/my-rule-2` already exists and you try to merge in a config with `foo/my-rule-2` that is a different object?

> In other words, I don't think this proposal currently offers a solution to the issues discussed in eslint/eslint#3458 where two shareable configs depend on different versions of a plugin.

You are correct, this design doesn't address this problem because I don't think it's important to solve. This is an extreme edge case that I think we've spent way too much time and energy trying to solve. With this design, configs can export a namespace for any rules they depend on through `ruledefs` such that `eslint-config-airbnb` could have a `ruledefs` with a `airbnbreact` namespace (or `airbnb::react` namespace, it's up to them). Shareable configs could either export two different namespaces, for example, both `airbnb::react` and `react`, both pointing to the same object containing rules.

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Reviewer 2: > That's a good point. Would it make sense to throw an error if `foo/my-rule-2` already exists and you try to merge in a config with `foo/my-rule-2` that is a different object?

Unfortunately this solution wouldn't be robust because it would be relying on the details of how a package manager flattens packages, which could unexpectedly change without action by the end user.

<details>
<summary>More explanation</summary>

Suppose `config1` depends on versions `^1.0.0` of `eslint-plugin-foo`, and `config2` depends on versions `~1.0.0` of `eslint-plugin-foo`. If the current version is `1.0.0` when the end user creates their config, and the end user uses a package manager that happens to flatten packages, then `eslint-plugin-foo@1.0.0` will get installed as a dependency of both, so the user's config will work. However, if `eslint-plugin-foo@1.1.0` is released later, it might end up that `eslint-plugin-foo@1.1.0` would get installed for `config1` and `eslint-plugin-foo@1.0.0` would get installed for `config2`. As a result, the end user's config would unexpectedly break later on without any action taken by the end user. Tree flattening in package managers is an implementation detail, so I don't think it's a good idea to have a solution that relies on two packages being resolved to the same object.

</details>

---

> You are correct, this design doesn't address this problem because I don't think it's important to solve. This is an extreme edge case that I think we've spent way too much time and energy trying to solve.

Broadly speaking, I think it's necessary to consider edge cases like this when designing a system of interaction between third parties. Given how many users we have, even rare edge cases will occur frequently in absolute terms.

The reason this problem keeps coming up is that we're asking third parties who don't have any information about each other to introduce things into a global namespace without colliding with each other. This is impossible in general.

To make the problem tractable, it's necessary to impose constraints on names that third parties can use, either by:

* Giving the end user final say over what names can be introduced (like our current `peerDependencies` solution, which allows the end user to install and choose the version of their plugins)
* Automatically assigning new names when third parties introduce a name conflict (like the solution in https://github.com/eslint/rfcs/pull/5, which assigns new names using namespaces)
* Letting config authors manually assign new names when two of their dependencies conflict with each other (like "Solution 1" from https://github.com/eslint/rfcs/pull/9#discussion_r251960102)

It's important to me that we have a workable way to address the problem rather than just hoping it doesn't occur in practice. I would rather have a robust system that's slightly more difficult to use than a simple system that unexpectedly breaks 2% of the time.

> Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Author: Just to be clear: I'm not saying I didn't consider this edge case, I'm saying I considered it and decided not to attempt to solve it directly. I don't believe solving this should be a requirement for this design, either.

As I've mentioned elsewhere, version conflicts between dependencies is really up to the package manager to solve. npm itself has complex logic to figure out which versions of which packages should be installed and in what directories to install them. If npm decides it cannot properly install two packages with different versions, then there's nothing ESLint can do to solve that.

There are also non-tech solutions to this problem. For example, if two configs conflict, the end user could manually copy what they wanted out of one shareable config into the place they want it. They could also manually install the plugin version that they want and stop using the shareable config altogether.

And just to set context, we have never seen this problem in the wild, we are only hypothesizing that this could happen (I believe I was the first to hypothesize this on https://github.com/eslint/eslint/issues/3458). I agree that doesn't mean the chance of it happening is 0%, but I do think the chance is sufficiently close to 0% that it's not something that should hold up this RFC.

Similarly, the `Config.assign()` method lets users modify their config by merging in changes from other configs, such as:

```js
const standardConfig = require("eslint-config-standard");
const personalConfig = require("@personal/eslint-config");
const { Config } = require("@eslint/config");

const myConfig = Config.create(standardConfig, {
    rules: {
        semi: "warn"
    }
});

Config.assign(myConfig, personalConfig);

exports.config = myConfig;
```

This code example merges `personalConfig` into `myConfig` using the same functionality that ESLint currently uses to merge configs.

#### The `RuleLoader` Class

The purpose of the `RuleLoader` class is aid in loading rules to replace `--rulesdir` functionality and has this form:

```js
class RuleLoader {

    // create a new instance
    constructor() {}

    // mimic `--rulesdir` loading

>     loadRulesFromDirectory(directory) {}

Author: I'm not a fan of exporting functions on their own. I'm also not sure `loadRulesFromDirectory()` will be the only method we end up needing on this class, so I'm more comfortable with having a class there.

>     loadRulesFromDirectory(directory) {}

Author: I think we just have different design philosophies on this point. My philosophy is that I don't like exporting functions on their own, and I don't like exporting namespace objects just to hang functions on. I much prefer creating a class to leave space for the possibility that there might be some stateful information necessary in the future. I'd prefer to keep this the way I have it defined and focus on larger points in this design.

>     loadRulesFromDirectory(directory) {}

Reviewer 2: Thanks for explaining. I disagree, but this isn't a hill I'm going to die on.

}
```

In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `ruledefs` and then specify the rules from a directory, such as:

```js
const { Config, RuleLoader } = require("@eslint/config");
const ruleLoader = new RuleLoader();

exports.config = new Config({
    ruledefs: {
        custom: ruleLoader.loadFromDirectory("./custom-rules")
    },
    rules: {
        "custom/my-rule": "error"
    }
});
```

The `RuleLoader#loadFromDirectory()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).

### Advanced Configs

> Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:

Reviewer 11: 👍 

Maybe it would also make sense to support async functions, that would be helpful for use-cases like shown in `ignore` example where the user wants to read a file.

For example:

```js
const fs = require("fs").promises;

exports.config = async () => {
    const [ file1, file2 ] = await Promise.all([ fs.readFile('./file1'), fs.readFile('./file2') ]);
    
    return {
        files: "*.js",
        ignore: `${file1}\n${file2}`.replace(/\n/g, ";")
    };
});
```

Reviewer 4: Can always be a async function?

Author: I don't believe we can do this as an async function *yet* because we would need to make a lot more of ESLint async to accomplish that. 

I do think we could add the ability to do async functions later on, though, as we figure out how to make more things async.

```js
exports.config = (context) => {

    // do something

    return {
        rules: {
            "semi": ["error", "always"]
        }
    };

};
```

The `context` object has the following members:

> * `core` - information about the ESLint core that is using the config

Author: I was purposely trying not to use "eslint" for this key in the off case that a config ends up in some ESLint wrapper like Standard or XO and to make it clear that this info was coming from the core that was running the config rather than the application itself (hence why you can't get plugin info, for example).

I'm not tied to `core`, though `eslintInfo` seems a bit verbose to me.

> * `core` - information about the ESLint core that is using the config

Reviewer 2: If a wrapper like Standard is running a config, is the expectation that the version property would represent the version of ESLint, or the version of Standard?

I had expected that it would represent the version of ESLint. For example, suppose ESLint v6.2.0 adds a particular rule option, and my config checks the `version` property to determine whether it should enable that option (based on whether the version is at least v6.2.0). If the `standard@4.1.0` package happens to depend on `eslint@6.3.0`, then passing `v4.1.0` as the `core.version` property doesn't seem like it would make much sense, because the version of Standard being used is largely irrelevant to what ESLint features are available.

As a result, including `eslint` in the key seems like it would make sense to me (or your aversion to including `eslint` in the key makes me suspect I'm misunderstanding something).

> * `core` - information about the ESLint core that is using the config

Author: The version would represent the ESLint core being used.

My idea here is that other tools might want to white label ESLint, even changing the name of the config file, so `eslint` might not necessarily mean anything to those users. They might also want to, for example, have other core rules (so `hasRule()`) could be different for a white label.

It's possible I'm overthinking it, though. I found naming this thing difficult. I wanted a way to say "this ESLint core has this version and contains these rules." I'm open to another name, but I'm not a fan of `eslintInfo`.

> * `core` - information about the ESLint core that is using the config

Reviewer 2: Thanks for clarifying. Given this reasoning I'm fine with keeping the name `core`.

    * `version` - the version of ESLint being used

>     * `hasRule(ruleId)` - determine if the given rule is in the core

Author: That's true. The intent was not to solve all possible version issues in configs, just to provide a little bit of context to help solve the most egregious issues.

    * `getConfig(configId)` - returns one of the predefined ESLint configs 
* `cwd` - the current working directory for ESLint (might be different than `process.cwd()`, see https://github.com/eslint/eslint/issues/11218)

This information allows users to make logical decisions about how the config should be constructed.

#### Checking for Rule Existence

> One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:

Reviewer 6: What about a new rule in the plugin? Would context automatically pick up all of the rules from plugins declared in `ruledefs`?

Reviewer 2: `context` is an argument to the config function, so `ruledefs` wouldn't exist at that point since it's generated by the config. Presumably, a config that has a plugin as a dependency would already be able to figure out which rules it has, since it's providing the rules at runtime.

Author: Correct.

};
```

#### Extending `eslint:recommended`

Advanced configs must be used to extend builtin configs such as `eslint:recommended` and `eslint:all`:

```js
const { Config } = require("@eslint/config");

exports.config = (context) => {

    const recommendedConfig = context.core.getConfig("eslint:recommended");

    return Config.create(recommendedConfig, {
        ruledefs: {
            custom: ruleLoader.loadFromDirectory("./custom-rules")
        },
        rules: {
            "custom/my-rule": "error"
        }
    });
};
```

### Configuration Location Resolution

When ESLint is executed, the following steps are taken to find the `.eslint.js` file to use:

1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
1. Otherwise:
    a. Look for `.eslint.js` in the current working directory. If founnd, stop searching and use that file.
    b. If not found, search up the directory hierarchy looking for `.eslint.js`.
    c. If a `.eslint.js` file is found at any point, stop searching and use that file.
    d. If `/` is reached without finding `.eslint.js`, then stop searching and output a "no configuration found" error.

> This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `.eslint.js` is found).

Reviewer 2: To clarify, this is assuming the `.eslint.js` file doesn't use the provided `cwd` value, right? Or would the CWD be set to the directory of the `.eslint.js` file by default?

Reviewer 2: I meant that the statement "This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory" is not necessarily true in the unusual case where a config file is explicitly changing what config it returns depending on the `cwd` argument, because the `cwd` would be different in the two invocations.

Author: Oh yes, good point. Yes, there is a theoretical way that could end up different but I'd expect that to be an edge case. So I think the word "allow" is correct in my statement rather than "enforce."

Some of the key differences from the way ESLint's configuration resolution works today are:

1. There is no automatic search for `.eslint.js` in the user's home directory. Users wanting this functionality can either pass a home directory file using `-c` or manually read in that file from their `.eslint.js` file.
1. Once a `.eslint.js` file is found, there is no more searching for any further config files.
1. There is no automatic merging of config files using either `extends` or `overrides`.
1. When `-c` is passed on the command line, there is no search performed.

### File Pattern Resolution

Because there are file patterns included in `.eslint.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:

1. When a filename is passed directly (such as `eslint foo.js`):
    1. ESLint checks to see if there is a config where the `files` pattern matches the file that was passed in. If no config is found, then the file is ignored (with an appropriate warning).

>     1. If a matching config is found, then the `ignore` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.

Reviewer 2: What if there are multiple matching configs for a particular file?

1. The `--print-config` option becomes less useful because we can't output objects in a meaningful way.
1. People depending on environments may find this change particularly painful.

## Backwards Compatibility Analysis

The intent of this proposal is to replace the current `.eslintrc` format, but can be implemented incrementally so as to cause as little disturbance to ESLint users as possible

In the first phase, I envision this:

1. `.eslint.js` can be implemented alongside `.eslintrc`.
1. If `.eslint.js` is found, then:
    1. All `.eslintrc` files are ignored.
    1. `.eslintignore` is ignored.
    1. `--ignore-file` is ignored.
    1. `--rulesdir` is ignored.
    1. `--env` is ignored.
    1. `eslint-env` config comments are ignored.
1. If `.eslint.js` is not found, then fall back to the current behavior.

This keeps the current behavior for the majority of users while allowing some users to test out the new functionality. Also, `-c` could not be used with `.eslint.js` in this phase.

In the second phase (and in a major release), ESLint will emit deprecation warnings whenever the original functionality is used but will still honor them so long as `.eslint.js` is not found.

> In the third phase (and in another major release), the original functionality is removed.

Reviewer 2: Minor: It might be worthwhile to have a fallback that looks for `.eslintrc.*` files and displays an informative error message about how they're no longer supported (possibly with a link to a migration guide or a codemod), rather than displaying the generic config-not-found message. This could ease migration confusion without too much added complexity.

Author: That's a fantastic idea. I'll add that in.

So while this is intended to be a breaking change, it will be introduced over the course of three major releases in order to give users ample time to transition.

> ## Alternatives

Reviewer 2: Bikesheddable: A name like `eslint.config.js` rather than `.eslint.js` might make it clearer that the file is a config file. It would also avoid the issue where the config file is ignored by default due to being a dotfile.

Author: Changed to `.eslint.config.js` per your and @kaicataldo's comments.

While there are no alternatives that cover all of the functionality in this RFC, there are alternatives designed to address various parts.

> * Both https://github.com/eslint/rfcs/pull/7 and https://github.com/eslint/rfcs/pull/5 specify an alternative method for resolving plugin locations in configs. This attempts to solve the problem with bundling plugins with configs (https://github.com/eslint/eslint/issues/3458). These proposals have the benefit of working within the current configuration system and requiring very few changes from users.

Reviewer 2: As a sidenote, given the major release timeline I still think it would be worth implementing https://github.com/eslint/rfcs/pull/7 regardless of whether we approve this RFC, to fix ESLint setups with Yarn Plug n' Play.

Author: I'm not opposed to that, and I didn't intend this RFC to block #7. I just wanted to make sure the problem was solved in this design too.

* It is possible to come up with a solution to https://github.com/eslint/eslint/issues/8813 that makes use of `extends`, though this would get a bit complicated if an extended config in an `overrides` section also has `overrides`.
* We could switch the current configuration system over so that all config files are considered to implicitly have `root:true`. This could dramatically simplify configuration searching and merging.

> ## Open Questions

Reviewer 2: Minor: The `--use-eslintrc` flag and the corresponding `CLIEngine` option would probably become confusing with this change. It might be a good idea to rename the flag.

Author: Good point. I'll add a note about that.

> 1. Should `ignore` allow line breaks so it can emulate `.eslintignore`?

Reviewer 2: I think it would be good for `ignore` to be an array rather than a string -- in my experience `.eslintignore` files usually have more than one line.

Author: Changed to an array.

> 1. Is `ruledefs` a create enough key name?

Author: Heh, I meant "clear". Not sure where "create" came from. :)

1. Do we need a command line flag to opt-in to `.eslint.js` instead of trying to do it alongside the existing configuration system?

> 1. Does the file pattern system actually remove the need for `--ext`?

Reviewer 2: How would this affect the `eslintConfig` and `eslintIgnore` `package.json` keys? Would they be deprecated/removed in the same manner as `.eslintrc.*` files?

Author: Yes. The intent is that all configuration now lives in just one place.

```js
config.setRuleSeverity("semi", "warn");
```

In this way, users don't need to modify an array to change a rule's severity, which is a major downside of this design when using a vanilla object structure as a config.

The static methods on `Config` make it easy to create and modify configs. For example, suppose you want to create a config based on `eslint-config-standard` but would prefer `semi` be a warning instead of an error, you can do so using `Config.create()`:

```js
const standardConfig = require("eslint-config-standard");
const { Config } = require("@eslint/config");

const myConfig = Config.create(standardConfig, {
    rules: {
        semi: "warn"
    }
});

exports.config = myConfig;
```

Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)

Similarly, the `Config.assign()` method lets users modify their config by merging in changes from other configs, such as:

> 

Reviewer 7: what happens if merging 2 files/ignores:

```js
var base = { ignores: ["dist/*", "coverage/*", ... ]};

var myConfig = Config.create(base, {ignores: [".*"]});  // ignores = ???
```

Author: Right now, `ignore` gets overwritten, which I think is the correct behavior. (You probably don't want a shareable config determining what your project does and doesn't ignore.)

Reviewer 8: You will if it’s not super easy to handle gitignore :-)

Design Summary:

1. Introduce a new `eslint.config.js` configuration file format
1. Searching for `eslint.config.js` starts from the current working directory and continues up
1. All `eslint.config.js` files are treated as if they have `root: true`
1. There is no automatic merging of config files
1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
1. Remove the concept of environments (`env`)
1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
1. Remove `--rulesdir`
1. Create a `@eslint/config` package to help users merge configs

### The `eslint.config.js` File

The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 

```js
exports.config = {
    files: ["*.js"],
    ignore: ["*.test.js"],
    extends: [],
    globals: {},
    settings: {},
    processor: object,
    parser: object,
    parserOptions: {},
    ruledefs: {},
    rules: {}
};
```

The following keys are new to the `eslint.config.js` format:

* `files` - **Required.** Determines the glob file patterns that this configuration applies to.

> * `ignore` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`.

Reviewer 8: the common case here is "gitignore", or "gitignore plus a few extras" - it'd be great if there was a way to let eslint pick this up automatically without having to import and split its contents here.

Author: I think letting users manually read in whatever file they want and process it however they want is the best approach here. It's easy to think that everyone follows a particular pattern, but I don't think we want to canonize any approach that involves files that ESLint doesn't own directly. (Perhaps there's room for a helper utility here, though I might want to wait for the first group of feedback from real users to decide what that looks like.)

Author: This is a very small corner of this proposal, so I'd suggest we table this discussion for now and revisit when the first implementation and batch of feedback has been collected to see if there is any complaints about this. For now, I think keeping this process manual sticks with the overall goals of this design.

Reviewer 8: As long as the possibility remains to add "use gitignore" in an ergonomic fashion, that seems totally reasonable.

exports.config = {
    files: ["*.js"],
    globals: {
        MyGlobal: true,
        ...globals.browser
    }
};
```

This effectively duplicates the use of `env: { browser: true }` in ESLint.

**Note:** This would allow us to stop environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.

#### Extending Another Config

Extending another config works the same as in `.eslintrc` except users pass entire config objects rather than the name of a config. That can easily be done in the following manner:

```js
exports.config = {
    files: ["*.js"],
    extends: require("eslint-config-standard"),
    rules: {
        semi: ["error", "always"]
    }
};
```

This config extends `eslint-config-standard` by assigning it to `extends`. You can also use an array for `extends` to extend from multiple configs:

```js
exports.config = {
    files: ["*.js"],
    extends: [
        require("eslint-config-standard"),
        require("@me/eslint-config")
    ],
    rules: {
        semi: ["error", "always"]
    }
};
```

#### Extending From `eslint:recommended` and `eslint:all`

> Both `eslint:recommended` and `eslint:all` can be represented as strings in the `extends` key. For example:

Author: Yeah, I wanted to avoid that for a couple of reasons:

1. I don't want configs to accidentally modify those values, therefore affecting how other configs in their `extends` chain are interpreted.
2. It creates a dependency mess between configs and ESLint. In this scenario, a config could require a different version of ESLint than what the user running. By allowing strings values, we avoid this dependency issue altogether and ensure that configs aren't requiring a certain version of ESLint work correctly (though they could check the version if they wanted to).

> Both `eslint:recommended` and `eslint:all` can be represented as strings in the `extends` key. For example:

Reviewer 12: The first I was contemplating could be a feature, so thanks for explaining your rationale there. The second I hadn't considered, and you're completely right - that would be gross to handle.

1. Create a `@eslint/config` package to help users manager configs

### The `eslint.config.js` File

The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 

```js
exports.config = {
    files: ["*.js"],
    ignores: ["*.test.js"],
    extends: [],
    globals: {},
    settings: {},
    processor: object,
    parser: object,
    parserOptions: {},
    ruledefs: {},
    rules: {}
};
```

The following keys are new to the `eslint.config.js` format:

* `files` - **Required.** Determines the glob file patterns that this configuration applies to.
* `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`.
* `ruledefs` - Contains definitions for rules grouped by a specific name. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.

The following keys are specified the same as in `.eslintrc` files:

* `globals`
* `settings`
* `rules`
* `parserOptions`

The following keys are specified differently than in `.eslintrc` files:

> * `extends` - an object or array in `eslint.config.js` files (a string or string array in `.eslintrc`)

Author: Oh wow, you are completely correct! By eliminating `extends`, this also greatly simplifies the merging logic. I'll work to get that incorporated. ❤️

> * `extends` - an object or array in `eslint.config.js` files (a string or string array in `.eslintrc`)

Author: `extends` has been removed. Thanks!

```yaml
plugins:
  - react

rules:
  react/jsx-uses-react: error
```

This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.

In `eslint.config.js`, the same configuration is achieved using a `ruledefs` key:

```js
const reactPlugin = require("eslint-plugin-react");

exports.config = {
    files: ["*.js"],
    ruledefs: {
        react: reactPlugin.rules
    },
    rules: {
        "react/jsx-uses-react": "error"
    }
};
```

Here, it is the `ruledefs` that assigns the name `react` to the rules from `eslint-plugin-react`. The reference to `react/` in a rule will always look up that value in the `ruledefs` key.

#### Referencing Parsers and Processors

In `.eslintrc`, the `parser` and `processor` keys required strings to be specified, such as:

```yaml
plugins: ["markdown"]
parser: "babel-eslint"
processor: "markdown/markdown"
```

In `eslint.config.js`, you would need to pass the references directly, such as:

```js
exports.config = {
    files: ["*.js"],
    parser: require("babel-eslint"),

>     processor: require("eslint-plugin-markdown").processors.markdown

Author: No. The final design for the `processor` key ended up being a single entry because after processing the result is matched against all possible `files` again. See https://github.com/eslint/rfcs/tree/master/designs/2018-processors-improvements#apply-processors-using-overrides

};
```

In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never a question of where the modules will be resolved from.

#### Applying an Environment

Unlike with `.eslintrc` files, there is no `env` key in `eslint.config.js`. Users can mimic the behavior of `env` by assigning directly to the `globals` key:

```js
const globals = require("globals");

exports.config = {
    files: ["*.js"],
    globals: {
        MyGlobal: true,
        ...globals.browser
    }

> };

Reviewer 8: even if this is a separate package, that'd be great

> };

Author: I'm not completely against the idea of including those globals in ESLint directly, but I'm not sure it has a ton of added value over just including the `globals` package. Ultimately, we'd just end up pulling in all of this from `globals` anyway rather than managing it ourselves. I'd like to hear some more opinions on this.

(Also, this is one of those things we can always add later if it turns out to be a pain point -- I don't know that it needs to be part of this design right now.)

> };

Reviewer: This is resolved by https://github.com/eslint/rfcs/pull/9/files#diff-4fab19e15273179c525fc49a689b73a5R447, though I think it's important to note that this will only work for the default parser.

> };

Author: This really isn't related parsers. Parsers don't know or care about globals, it's only ESLint that cares. By moving the `ecmaVersion` to a language-level property, ESLint (not the parser) will be in charge of defining globals so even custom parsers will benefit.

```yaml
plugins: ["react"]
rules:
   react/jsx-uses-react: error
    semi: error

overrides:
    - files: "*.md"
      plugins: ["markdown"],
      processor: "markdown/markdown"
```

This can be written in `eslint.config.js` as an array of two configs:

```js
exports.config = [
    {
        files: "*.js",
        ruledefs: {
            react: require("eslint-plugin-react").rules,
        },
        rules: {
            "react/jsx-uses-react": "error",
            semi: "error"
        }
    },
    {
        files: "*.md",
        processor: require("eslint-plugin-markdown").processors.markdown
    }
];
```

When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.

> If a config in the config array does not contain `files` or `ignores`, then that config applies to all files. For example:

Author: Yes, the config returned by `exports.config` must have a `files` property in order to match against the files that are passed on the command line. However, any single object inside of a config array need not have `files` so that it applies to everything. I'll add some text to make that clear.

```js
exports.config = [
    {
        globals: {
            Foo: true
        }
    },
    {
        files: "*.js",
        ruledefs: {
            react: require("eslint-plugin-react").rules,
        },
        rules: {
            "react/jsx-uses-react": "error",
            semi: "error"
        }
    },
    {
        files: "*.md",
        processor: require("eslint-plugin-markdown").processors.markdown
    }

> ];

Reviewer 2: I think the "config applies to all files by default" behavior is potentially confusing when combined with the removal of the `--ext` option.

Suppose I'm using the config given here, and I'm linting a `src/` folder containing `foo.js`, `README.md`, and `index.html`. If I'm understanding the file resolution section correctly, adding `globals: { Foo: true }` to the config array would mean that there is now a matching entry for `index.html` (since the new entry matches all files), so `index.html` would suddenly start getting linted (probably resulting in a parsing error).

Author: There is no config with a `files` pattern matching `*.html`, so `index.html` will not get linted. A config without `files` applies to all files that are passed in so long as there is a config with a matching `files` pattern. A config without a `files` pattern is never matched on its own. I'll add some text to clarify that.

```

In this example, the first config in the array defines a global variable of `Foo`. That global variable is merged into the other two configs in the array automatically because there is no `files` or `ignores` specifying when it should be used.

Each item in a config array can be a config array. For example, this is a valid config array:

```js
exports.config = [
    [
        require("eslint-config-vue"),
        require("eslint-config-import")
    ]
    {
        files: "*.js",
        ruledefs: {
            react: require("eslint-plugin-react").rules,
        },
        rules: {
            "react/jsx-uses-react": "error",
            semi: "error"
        }
    },
    {
        files: "*.md",
        processor: require("eslint-plugin-markdown").processors.markdown
    }
];
```

> A config array is always flattened before being evaluating, so even though this example is a two-dimensional config array, it will be evaluated as if it were a one-dimensional config array.

Reviewer 2: ```suggestion
A config array is always flattened before being evaluated, so even though this example is a two-dimensional config array, it will be evaluated as if it were a one-dimensional config array.
```

> A config array is always flattened before being evaluating, so either though this example is a two-dimensional config array, it will be evaluated as if it were a one-dimensional config array.

Reviewer 12: ```suggestion
A config array is always flattened before being evaluating, so even though this example is a two-dimensional config array, it will be evaluated as if it were a one-dimensional config array.
```

#### The `RuleLoader` Class

The purpose of the `RuleLoader` class is to aid in loading rules to replace `--rulesdir` functionality and has this form:

```js
class RuleLoader {

    // create a new instance
    constructor() {}

    // mimic `--rulesdir` loading
    loadRulesFromDirectory(directory) {}
}
```

In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `ruledefs` and then specify the rules from a directory, such as:

```js
const { RuleLoader } = require("@eslint/config");
const ruleLoader = new RuleLoader();

exports.config = {
    ruledefs: {
        custom: ruleLoader.loadFromDirectory("./custom-rules")
    },
    rules: {
        "custom/my-rule": "error"
    }
};
```

The `RuleLoader#loadFromDirectory()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).

### Advanced Configs

> Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:

Author: 1. The function must return an object or an array.
2. No.
3. No.

It helps to think of the function like an AWS Lambda function. Inside of that function you can do whatever you want, but returning a function from that function won't result in a second function call.

I'll add in some details about this.

> Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:

Reviewer 2: Does this mean that in order to use a shareable config that exports a function, an end user would also be required to make their config a function (so they could call the shareable config with the right arguments)?

I'm concerned that this would cause some problems. For example:

* A shareable config would be required to make a breaking change to convert their config from an object/array to a function.
* An end user might be confused about why some shareable configs can be extended with `extends: [require('eslint-config-foo')]`, and other shareable configs need to be extended with `extends: [require('eslint-config-foo')(context)]`.

Could you elaborate on what we gain by disallowing functions in `extends`? In my opinion, it seems better to allow a function anywhere we allow an object config.

> Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:

Author: > Does this mean that in order to use a shareable config that exports a function, an end user would also be required to make their config a function (so they could call the shareable config with the right arguments)?

That was my intention, yes. However, you make some good points and have convinced me that it makes sense to allow a function anywhere a config object is currently allowed. I'll work on adding that into this design this week.

    d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.

This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `eslint.config.js` is found).

Some of the key differences from the way ESLint's configuration resolution works today are:

1. There is no automatic search for `eslint.config.js` in the user's home directory. Users wanting this functionality can either pass a home directory file using `-c` or manually read in that file from their `eslint.config.js` file.
1. Once a `eslint.config.js` file is found, there is no more searching for any further config files.
1. There is no automatic merging of config files using either `extends` or `overrides`.
1. When `-c` is passed on the command line, there is no search performed.

### File Pattern Resolution

Because there are file patterns included in `eslint.config.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:

1. When a filename is passed directly (such as `eslint foo.js`):

>     1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern. All configs with a matching `files` pattern are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.

Author: Yeah, that's what I meant to happen. I can see that's not very clear so I'll update this section.

>     1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.

Reviewer 2: How does this affect default ignore patterns such as `node_modules`? I'm assuming we would still ignore `node_modules` by default when running `eslint .`, but would `node_modules/foo.js` still be ignored if it was explicitly listed in a `files` pattern?

Author: That's a good question. I was assuming that people would manually specify to ignore `node_modules`, but we probably do want to ignore that by default as we do now. I'll add in a note about that.

>     1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.

Reviewer 2: Could you elaborate on how `files` and `ignores` are merged with `extends`? 

```js
exports.config = {
    files: ["*.js", "*.ts"],
    extends: [
        {
            files: ["*.ts"]
            ...extendedConfig
        }
    ],
    ...baseConfig
};
```

If I'm understanding the discussion at https://github.com/eslint/rfcs/pull/9#discussion_r251180712 correctly, would that config be equivalent to this?

```js
exports.config = [
    {
        files: ["*.ts"]
        ...extendedConfig
    },
    {
        files: ["*.js", "*.ts"],
        ...baseConfig
    }
];
```

Author: This is one of the open questions I listed in the design. My initial thought was that `files` and `ignores` would be overwritten rather than merged.

However, if we remove `extends` (as @mysticatea has suggested), then I think this becomes a lot clearer.

Reviewer 2: Another option would be to make `extends` be syntax sugar for array configs, i.e.

```js
module.exports = {
    ...config,
    extends: [shareableConfig]
}
```

...would always be equivalent to:

```js
module.exports = [
    shareableConfig,
    config
]
```

In most cases, I think shareable configs probably wouldn't specify file patterns anyway, and having `extends` might be more familiar/readable to new users (who otherwise might not be sure what array configs do). I don't feel strongly about this, however.

Author: Please take a look at the latest update. I think removing `extends` actually makes things a lot simpler. Basically, a config array acts just like an `extends` array does now, so I think it will feel pretty familiar to users.

    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
1. When a glob pattern is passed directly (such as `eslint src/*.js`):
    1. ESLint expands the glob pattern to get a list of files.
    1. Each file is checked individually as in step 1.

> 1. When a directory is passed directly (such as `eslint src`):

Reviewer 6: What happens when eslint is executed as `eslint .`? Also, since config now contains globs, should we allow just execution of the eslint without any parameters (so that files to be linted would be resolved from the config globs?).

Author: When `eslint .` is run, the `.` is resolved to the current working directory. I can add another step to make that clear.

Before I posted this RFC, my thought was that we could run ESLint without specifying any files or directories, but I removed it because that seemed like it could be confusing. I'm open to adding it back in if it seems useful.

Reviewer 6: If you could. I understand that `.` will just be a glob for current directory, but it's a widely used way of running ESLint, and I think it would be clearer if we mention it explicitly.

In terms of running ESLint without anything, I think it makes sense, given that we can configure globs in the config now.

Author: I added a step to clarify what `eslint .` does.

I'm going to add your question about running `eslint` without any files into the Open Questions section to get some more feedback.

>     1. The directory is converted into a glob pattern by appending the contents of the `files` array (such as `src` becomes `src/**/*.js`).

Author: Yeah you're right. I need to think more about these implementation details. In the meantime, I think adding `**/*` to the end of each directory makes sense (we can always work on improving performance during implementation as well).

## Alternatives

While there are no alternatives that cover all of the functionality in this RFC, there are alternatives designed to address various parts.

* Both https://github.com/eslint/rfcs/pull/7 and https://github.com/eslint/rfcs/pull/5 specify an alternative method for resolving plugin locations in configs. This attempts to solve the problem with bundling plugins with configs (https://github.com/eslint/eslint/issues/3458). These proposals have the benefit of working within the current configuration system and requiring very few changes from users.
* It is possible to come up with a solution to https://github.com/eslint/eslint/issues/8813 that makes use of `extends`, though this would get a bit complicated if an extended config in an `overrides` section also has `overrides`.
* We could switch the current configuration system over so that all config files are considered to implicitly have `root:true`. This could dramatically simplify configuration searching and merging.

## Open Questions

1. Is `ruledefs` a clear enough key name?

> 1. Do we need a command line flag to opt-in to `eslint.config.js` instead of trying to do it alongside the existing configuration system?

Author: :+1: And again, this would just be during the initial implementation, so I don't think we need to support every corner case right out the gate.

> 1. Do we need a command line flag to opt-in to `eslint.config.js` instead of trying to do it alongside the existing configuration system?

Reviewer: Agreed. I don't think a flag is necessary.

### Is there a way to allow plugins to specify the namespace of their rules?

Maybe, but even if we could do that, the namespaces aren't guaranteed to be unique.

The reason we could enforce naming of both plugin packages and rule namespaces is because ESLint controlled how the plugins were loaded: users passed ESLint a string and then ESLint could both inspect the string to pull out details it needed (the plugin name without the `eslint-plugin-` prefix) and then modify the rule names to be prepended with the plugin name.

In this design, the user is responsible for loading npm packages. Because we are only ever passed an object, we no longer have access to the plugin name.

An earlier iteration of this design had plugins specified like this:

```js
exports.config = {
    plugins: [
        require("eslint-plugin-react"),
        require("eslint-plugin-mardkown")
    ]
};
```

> This didn't work because ESLint didn't know how to name the plugin rules (we weren't getting the plugin name, just the object). So the next iteration went to this:

Reviewer 6: Can we enforce a specific format on the exported object for plugin creators, something that contains property `name` for example? It doesn't solve the issue of collision, but would not allow renaming of the plugins.

Author: I'm adding a new section about this. I'd like to see if it solves your concern.

## Detailed Design

Design Summary:

1. Introduce a new `eslint.config.js` configuration file format
1. Searching for `eslint.config.js` starts from the current working directory and continues up
1. All `eslint.config.js` files are treated as if they have `root: true`
1. There is no automatic merging of config files
1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
1. Remove the concept of environments (`env`)
1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
1. Remove `--rulesdir`
1. Remove `--ext`

### The `eslint.config.js` File

The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 

```js
exports.config = {
    name: "name",
    files: ["*.js"],
    ignores: ["*.test.js"],
    globals: {},
    settings: {},
    processor: object,
    parser: object,
    parserOptions: {},

>     ruledefs: {},

Author: Remember what we said about bikeshed comments? :)

I'd prefer to keep it as "ruledefs" because I think of it as similar to "typedef"

>     ruledefs: {},

Reviewer 2: Ok, I'm fine with keeping "ruledefs".

Aside: I think we decided to stop making "nitpick" (non-blocking) comments, not bikeshed comments. The names here are a substantial part of the UX and I think discussion about them is worthwhile.

>     ruledefs: {},

Reviewer 13: I personally prefer `ruleDefs`.

```
your-project
├── eslint.config.js
└── node_modules
  ├── eslint
  ├── eslint-config-a
  | └── node_modules
  |   └── eslint-plugin-example@1.0.0
  └── eslint-config-b
    └── node_modules
      └── eslint-plugin-example@2.0.0
```

The problem comes when the shareable configs try to use the default namespace of `eslint-plugin-example` for its rules, such as:

```js
module.exports = {
    ruledefs: {
        ...require("eslint-plugin-example").ruledefs
    }
};
```

> If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown because the `ruledefs` namespace "example" can only be assigned once.

Reviewer 2: To be clear, is an error thrown when two shareable configs with conflicting ruledefs are *merged*, or simply whenever they both appear in a final config at all (even if they apply to disjoint sets of files)?

Based on https://github.com/eslint/rfcs/pull/10#issuecomment-460706972 it seems like it's the latter, but it might be worthwhile to clarify this.

Author: It will happen when they are merged, I'll update that.

Reviewer 2: In https://github.com/eslint/rfcs/pull/10#issuecomment-460706972 you mentioned that rule IDs are unique within a particular config. If conflict detection only happens when configs are merged, it seems like rule IDs wouldn't be unique for this config:

```js
exports.config = [
  {
    files: ["*.js"],
    ruledefs: {
      foo: RULESET_ONE
    }
  },
  {
    files: ["*.ts"],
    ruledefs: {
      foo: RULESET_TWO
    }
  }
]
```

There are no files that match both configs, so the two configs would never be merged and no conflict would be detected. However, there would be a duplicate "foo" ruledef, and possibly duplicate rule IDs, which would cause problems for https://github.com/eslint/rfcs/pull/10. I think it would be better for an error to be thrown for this case.

Author: That's a good point, we can certainly detect that during the normalization phase.

> If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown because the `ruledefs` namespace "example" can only be assigned once.

Author: > For example, if the `eslint-plugin-node` plugin exports two configs, both of which use `eslint-plugin-node`'s rules, then the user won't be able to extend both of them because they will both have a `node` ruledef containing `eslint-plugin-node`'s rules, even though both configs were created by the same author and intended to share the same rules.

Do you know of an example where this happens today?

My opinion is that it would be up to the plugin to provide documentation about the correct way to use its exported configs.

> If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown because the `ruledefs` namespace "example" can only be assigned once.

Reviewer 2: > Do you know of an example where this happens today?

Yes, I think it's pretty common. For example, [`eslint-plugin-eslint-plugin`](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/9bab97423656bb40a4cd199034c5a167e89b822e/lib/index.js) exports `rules` and `tests` configs which enable different subsets of the plugin's rules.

> To work around this problem, shareable configs that rely on plugins should also export a separate config that assigns a unique namespace to all plugin rules. To continue with the example in this section, that might be under `eslint-config-example/compat` and would look like this:

Author: That's fine, we can always decide on what the best way to approach this is when we write the user-facing documentation. My point here is just to show that it's possible.

> To work around this problem, shareable configs that rely on plugins should also export a separate config that assigns a unique namespace to all plugin rules. To continue with the example in this section, that might be under `eslint-config-example/compat` and would look like this:

Reviewer 2: In that case, can we tweak the wording here? It currently says "shareable configs that rely on plugins should also export a separate config that assigns a unique namespace to all plugin rules", which seems like it's already recommending a particular approach.

> To work around this problem, shareable configs that rely on plugins should also export a separate config that assigns a unique namespace to all plugin rules. To continue with the example in this section, that might be under `eslint-config-example/compat` and would look like this:

Author: That's fine.

```suggestion
To work around this problem, shareable configs that rely on plugins may also export a separate config that assigns a unique namespace to all plugin rules. To continue with the example in this section, that might be under `eslint-config-example/compat` and would look like this:
```

for (const key in originalRuleDefs) {
    compatRuleDefs["config-a::" + key] = originalRuleDefs[key];
}

// include in a config
module.exports = {
    ruledefs: {
        ...compatRuleDefs
    }
};
```

Here, the shareable config is exporting the same rules from the plugins as before, but using the namespace `config-a::example` instead of the default `example` that `eslint-plugin-example` defines. In this way, shareable config authors can provide an option for users who may be extending multiple configs that depend on the same plugin.

If a shareable config does not provide a compat version, then the end user still can create one on their own by manually creating a new config from the shareable config. For example:

```js
// get the config you want to extend
const configToExtend = require("eslint-plugin-example");

// create a new copy
const compatConfig = Object.create(configToExtend);
compatConfig.ruledefs = {};

// namespace each original
for (const key in configToExtend.ruledefs) {
    compatConfig.ruledefs["compat::" + key] = configToExtend.ruledefs[key];
}

// include in config
exports.config = [
    compatConfig,
    {
        // overrides here
    }
];
```

> 

Reviewer 2: Could we document which of the following changes would constitute breaking changes for a shareable config, taking into account the possibility of causing errors from ruledef conflicts?

1. Adding a new plugin (ruledef) to a shareable config, without enabling any of its rules
1. Configuring a rule from an existing ruledef which previously was not configured
1. Adding more rules to an existing ruledef (e.g. by upgrading a plugin), without configuring the rules

My understanding is that (1) would be a breaking change because it could cause a new ruledef conflict for the config's users. (2) would be a breaking change for other reasons (it would usually cause more errors to be reported for end users). (3) depends on whether we decide to allow "conflicting" ruledefs that have the same object identity.

Author: Can we simplify this by saying that any change to a shareable config's `ruledefs` is a breaking change?

Reviewer 2: Possibly, although making (3) a breaking change seems undesirable (e.g. it would include bugfixes in plugins). Unless a shareable config pinned its plugin dependencies to an exact version, a "breaking change" like this could occur accidentally if a new version of a plugin is released. This is a disadvantage of having behavior changes dependent on how package managers flatten packages.

Author: I don't want to get too far into prescribing when shareable configs should consider a new release a breaking change (we don't do this today and arguably already have the same problem). I'd rather just warn shareable config authors that *any* change, whether that be to a dependency or to what they're outputting is possibly a breaking change and leave it at that. (Arguably absolutely any change will affect people's builds.)

1. When a relative directory is passed directly (such as `eslint .`):
    1. The relative directory pattern is resolved to a full directory name.
    1. The glob pattern is checked as in step 3.

**Note:** ESLint will continue to ignore `node_modules` by default.

### Rename `--no-eslintrc` to `--no-config-file` and `useEslintrc` to `useConfigFile`

Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.

## Documentation

This will require extensive documentation changes and an introductory blog post.

At a minimum, these pages will have to be updated (and rewritten):

* https://eslint.org/docs/user-guide/getting-started#configuration
* https://eslint.org/docs/user-guide/configuring
* https://eslint.org/docs/developer-guide/working-with-plugins#configs-in-plugins
* https://eslint.org/docs/developer-guide/shareable-configs

## Drawbacks

As with any significant change, there are some significant drawbacks:

> 1. We'd need to do a phased rollout (see Backwards Compatibility Analysis) to minimize impact on users. That means maintaining two configuration systems for a while, increasing maintenance overhead and complexity of the application.

Author: I'm not sure having a codemod is possible. Given that people could be using any of four different formats for their current config file, plus mixing in `extends` and `overrides`, I think it would be difficult to go in and modify those configs and make them work.

I initially thought we might be able to provide a function that could at least convert existing configs after they are read in, but once again, the complexities of merging, module resolving, etc., makes that appear impractical.

Overall, I don't think the format is different enough that it would be difficult for people to convert themselves.

> 1. We'd need to do a phased rollout (see Backwards Compatibility Analysis) to minimize impact on users. That means maintaining two configuration systems for a while, increasing maintenance overhead and complexity of the application.

Reviewer 2: I'm not sure I understand the difficulty of doing this automatically. Could you clarify?

* It seems trivial for a tool to look at a project and determine which of the four config formats a project is using (it could just check for the existence of those files in the filesystem, like ESLint currently does).
* `extends: foo` can be transformed within a single config to put `require('eslint-config-foo')` in a config array.
* A similar transformation applies to `overrides`:

    ```js
    // old config
    module.exports = {
        ...baseConfig,
       overrides: {
           files: ["foo"],
           ...overrideConfig
       },
    };
    ```

    ```js
    // equivalent new config
    exports.config = [
        baseConfig,
        {
            files: ["foo"],
            ...overrideConfig
        }
    ];
    ```

* A runtime conversion tool also doesn't seem very complicated because we already have code that handles merging/module resolving (namely, the current config logic within ESLint).

I agree that the format isn't very difficult for a human to convert if they know what they're doing, but I think most of our users aren't very familiar with how configs work, and making everyone learn the new config format is a very large amount of migration pain.

Doing a back-of-the-envelope calculation:

* We have about 30 million downloads per week on npm. Assuming an average active project downloads ESLint about 30 times per week, there are about a million projects with old-style config files.
* Supposing it takes an average user 20 minutes to read our new config documentation and understand how to fix their config, that's about 333333 person-hours, or 38 person-years, spent migrating configs from the old format to the new format.

These are very rough estimates and could easily be off by a factor of four in either direction, but my point is that making our existing users learn our new config format would create a massive amount of work for them. Creating a tool to do the migration automatically would be a huge time-saver, and I think it's worth doing even if working out the details takes some time.

> 1. We'd need to do a phased rollout (see Backwards Compatibility Analysis) to minimize impact on users. That means maintaining two configuration systems for a while, increasing maintenance overhead and complexity of the application.

Reviewer 14: @not-an-aardvark this is already handled by using `@eslint/config` utilities? I don't see anything more specific to do.

1. Getting everyone to convert to `eslint.config.js` format would be a significant stress on the community that could cause some resentment.
1. We can no longer enforce naming conventions for plugins and shareable configs. This may make it more difficult for users to find compatible npm packages.
1. Creating configuration files will be more complicated.
1. The `--print-config` option becomes less useful because we can't output objects in a meaningful way.
1. People depending on environments may find this change particularly painful.

## Backwards Compatibility Analysis

The intent of this proposal is to replace the current `.eslintrc` format, but can be implemented incrementally so as to cause as little disturbance to ESLint users as possible

> In the first phase, I envision this:

Author: I don't think labeling it as experimental is helpful and might work counter to what we need by dissuading people from trying it out. If we agree that this is a good design to implement, then we should commit to it. The idea of phasing this format in and phasing out the old format is to catch any usability or other issues and address them, just like any other feature. It's unlikely people will jump immediately to the new format unless it's solving an immediate need, so we should have plenty of time before we remove the old functionality.

> In the first phase, I envision this:

Reviewer 2: Agreed that the goal is to catch usability issues, and that we should commit to this design before implementing it. I'm saying that if fixing a usability issue requires a breaking change to the new format, it could complicate adoption given our typical scheduling for major releases. By marking the new config format as experimental, we would be able to fix usability issues without waiting a year for the next major release, and then requiring users who migrated early to migrate again.

> In the first phase, I envision this:

Author: Again, I just don't see this as being any different from any other feature we implement.

> In the first phase, I envision this:

Reviewer 14: @not-an-aardvark, I don't think we need any "experimental" thing. It's different enough. It's opt-in anyway. If you choose to use the `eslint.config.js` then you know what you're doing and that it's experimental. We don't need to have "--experimental" for or such things. Or what you mean by "experimental" more specifically?

1. `eslint.config.js` can be implemented alongside `.eslintrc`.
1. If `eslint.config.js` is found, then:
    1. All `.eslintrc` files are ignored.

>     1. `.eslintignore` is ignored.

Author: That seems like a good idea. I'll add that in. Thanks!

    1. `--ignore-file` is ignored.
    1. `--rulesdir` is ignored.
    1. `--env` is ignored.
    1. `--ext` is ignored.

>     1. `--use-eslintrc` is ignored.

Author: This is a typo. There is no `--use-eslintrc`, there's just `--no-eslintrc`. I'll remove that.

### Why use `exports.config` instead of `module.exports`?

Using an exported key gives us more flexibility for the future if we decide that config files should be able to output more than one thing. For example, I've been thinking of a `--config-key` option that would allow users to specify which exported key should be used as their config. Users could then export multiple different keys (`config1`, `config2`, etc.) and easily switch between configs on the command line. That option is not part of this proposal because it isn't solving an existing problem and I'd rather focus on existing problems first (this proposal is already big enough).

### How does this affect configuration via `package.json`?

The `eslintConfig` and `eslintIgnore` keys in `package.json` will not be honored when `eslint.config.js` is found. Users could still pull that information into their `eslint.config.js` file manually if they want to.

### Do shareable configs still export an object on `module.exports`?

> That is completely up to the shareable config. For simplicity sake, I think we still want to encourage people to do so. However, there is no longer and formal contract between ESLint and shareable configs, so developers could potentially export configs from any npm package using any exported key. They would just need to inform users about how to extend their config properly.

Reviewer 9: ```suggestion
That is completely up to the shareable config. For simplicity sake, I think we still want to encourage people to do so. However, there is no longer a formal contract between ESLint and shareable configs, so developers could potentially export configs from any npm package using any exported key. They would just need to inform users about how to extend their config properly.
```

### Why are "eslint:recommended" and "eslint:all" strings instead of objects?

> In order to use objects for these two configs, we'd need to somehow pass those objects to the config file. That would mean either exposing something on the ESLint package itself (i.e., `require("eslint").configs.recommended`) or publishing a separate package to do the same (i.e., `require("@eslint/configs).recommended). 

Reviewer 9: ```suggestion
In order to use objects for these two configs, we'd need to somehow pass those objects to the config file. That would mean either exposing something on the ESLint package itself (i.e., `require("eslint").configs.recommended`) or publishing a separate package to do the same (i.e., `require("@eslint/configs).recommended`). 
```

1. All `eslint.config.js` files are treated as if they have `root: true`
1. There is no automatic merging of config files
1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
1. Remove the concept of environments (`env`)
1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
1. Remove `--rulesdir`
1. Remove `--ext`

### The `eslint.config.js` File

The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 

```js
exports.config = {
    name: "name",
    files: ["*.js"],
    ignores: ["*.test.js"],
    globals: {},
    settings: {},
    processor: object,
    parser: object,
    parserOptions: {},
    defs: {
        ruleNamespaces: {}
    },
    rules: {}
};
```

The following keys are new to the `eslint.config.js` format:

* `name` - Specifies the name of the config object. This is helpful for printing out debugging information and, while not required, is recommended for that reason.
* `files` - **Required.** Determines the glob file patterns that this configuration applies to.

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`.

Reviewer: Just to clarify, would this still support negated patterns (such as `!.*.js`)?

#### Extending Another Config

Extending another config is accomplished by returning an array as the value of `exports.config`. Configs that come later in the array are merged with configs that come earlier in the array. For example:

```js
exports.config = [
    require("eslint-config-standard"),
    {
        files: ["*.js"],
        rules: {
            semi: ["error", "always"]
        }
    }
];
```

This config extends `eslint-config-standard` because that package is included first in the array. You can add multiple configs into the array to extend from multiple configs, such as:

```js
exports.config = [
    require("eslint-config-standard"),
    require("@me/eslint-config"),
    {
        files: ["*.js"],
        rules: {
            semi: ["error", "always"]
        }
    }
];
```

Each item in a config array can be a config array. For example, this is a valid config array and equivalent to the previous example:

```js
exports.config = [
    [
        require("eslint-config-standard"),
        require("@me/eslint-config")
    ],
    {
        files: ["*.js"],
        rules: {
            semi: ["error", "always"]
        }
    }
];
```

> A config array is always flattened before being evaluated, so even though this example is a two-dimensional config array, it will be evaluated as if it were a one-dimensional config array.

Reviewer 13: Out of curiosity, how deeply can config arrays be nested?

> When using a config array, only one config object must have a `files` key (config arrays where no objects contain `files` will result in an error). If a config in the config array does not contain `files` or `ignores`, then that config is merged into every config with a `files` pattern. For example:

Reviewer 13: ```suggestion
When using a config array, at least one config object must have a `files` key (config arrays where no objects contain `files` will result in an error). If a config in the config array does not contain `files` or `ignores`, then that config is merged into every config with a `files` pattern. For example:
```

Reviewer 13: The current wording implies that only one occurrence of&nbsp;the&nbsp;`files`&nbsp;key is&nbsp;allowed in&nbsp;a&nbsp;config to&nbsp;me.

Reviewer 15: @kaicataldo well in English "only" and "at least" are totally different I guess ^^ Hope this clarify the confusion

> When using a config array, only one config object must have a `files` key (config arrays where no objects contain `files` will result in an error). If a config in the config array does not contain `files` or `ignores`, then that config is merged into every config with a `files` pattern. For example:

Reviewer: I wonder if `ignores` in shared configs or plugins should be allowed. My concern is that a plugin or shared config's `ignores` could be updated and the linting task could begin silently ignoring files the user is expecting to be linted.

Reviewer: I've changed my mind on this, as I think the benefits of having a shared `ignores` key in a base config outweighs the potential downside of things silently being ignored (though this is still a real danger).

To work around this problem, the end user can create a separate namespace for the same plugin so that it doesn't conflict with an existing rule namespace from a shareable config. For example, suppose you want to use `eslint-config-first`, and that has an `example` rule namespace defined. You'd also like to use `eslint-config-second`, which also has an `example` rule namespace defined. Trying to use both shareable configs will throw an error because a rule namespace cannot be defined twice. You can still use both shareable configs by creating a new config from `eslint-config-second` that uses a different namespace. For example:

```js
// get the config you want to extend
const configToExtend = require("eslint-config-second");

// create a new copy (NOTE: probably best to do this with @eslint/config somehow)
const compatConfig = Object.create(configToExtend);
compatConfig.defs.ruleNamespaces = Object.assign({}, configToExtend.defs.ruleNamespaces);
compatConfig.defs.ruleNamespaces["compat::example"] = require("eslint-plugin-example").rules;
delete compatConfig.defs.ruleNamespaces.example;

// include in config
exports.config = [

>     require("eslint-config-first");

Reviewer 9: ```suggestion
    require("eslint-config-first"),
```

**Note:** This would allow us to stop shipping environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.

#### Overriding Configuration Based on File Patterns

Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `eslint.config.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:

```yaml
plugins: ["react"]
rules:
    react/jsx-uses-react: error
    semi: error

overrides:
    - files: "*.md"
      plugins: ["markdown"],
      processor: "markdown/markdown"
```

This can be written in `eslint.config.js` as an array of two configs:

```js
exports.config = [
    {
        files: "*.js",
        defs: {
            ruleNamespaces: {
                react: require("eslint-plugin-react").rules,
            },
        }
        rules: {
            "react/jsx-uses-react": "error",
            semi: "error"
        }
    },
    {
        files: "*.md",
        processor: require("eslint-plugin-markdown").processors.markdown
    }
];
```

> When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.

Reviewer 16: How do this reconcile differences between files patterns in shared configs?  For example, let's say I pull in a shared config that specifies `files: "*.js"`, as a developer, would I understand why those rules weren't also applying to my `.jsx` files as well?  Should there be guidance to config developers about whether or not they should have `files` in their configs?  With the example above, with both markdown and js sections, which section do rules without explicit `files` get applied?

Author: In general, I think most shared configs will either a) not specify `files` or b) if they do specify `files`, indicate that so in their README. 

So yes, I think we'd want to provide guidance to config developers to make sure they specify which files their configs will apply to by default.

* `cwd` - the current working directory for ESLint (might be different than `process.cwd()` but always matches `CLIEngine.options.cwd`, see https://github.com/eslint/eslint/issues/11218)

This information allows users to make logical decisions about how the config should be constructed.

A configuration function may return an object or an array of objects. An error is thrown if any other type of value is returned.

#### Checking for Rule Existence

One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:

```js
const requireIndex = require("requireindex");

exports.config = (context) => {
    const myConfig = {
        files: ["*.js"],
        defs: {
            ruleNamespaces: {
                custom: requireIndex("./custom-rules")
            },
        }
        rules: {
            "custom/my-rule": "error"
        }
    };

>     if (context.hasRule("some-new-rule")) {

Reviewer 16: Could this be simplified by filtering out falsy rules?  e.g.:
```
    rules: {
        "some-new-rule": context.hasRule("some-new-rule") ? ["error"] : null
    }
```
Or, perhaps a helper function:  `context.ruleIfExists('some-new-rule, ["error"])`

Author: I feel like we won't really know what helper functions are useful until we get a base implementation out. It'll be easy to see what's missing once we get to see the feedback, whereas we can never remove anything if we add something that isn't needed.

```js
exports.config = [
    (context) => someObject,
    require("eslint-config-myconfig")
];
```

Each function config in an array will be executed with a `context` object when ESLint evaluates the configuration file. This also means that shareable configs can export a function instead of an object or array.

**Note:** If a function config inside of a config array happens to return an array, then those config array items are flattened as with any array-in-array situation.

### Configuration Location Resolution

When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:

1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
1. Otherwise:
    a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
    b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
    c. If a `eslint.config.js` file is found at any point, stop searching and use that file.
    d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.

> This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `eslint.config.js` is found).

Reviewer: This seems like a nice solution to the issue of always having to have the root directory of a repository open as the root in the file tree for editor integrations to work.

    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
1. When a glob pattern is passed directly (such as `eslint src/*.js`):
    1. ESLint expands the glob pattern to get a list of files.
    1. Each file is checked individually as in step 1.
1. When a directory is passed directly (such as `eslint src`):
    1. The directory is converted into a glob pattern by appending `**/*` to the directory (such as `src` becomes `src/**/*`).
    1. The glob pattern is checked as in step 2.
1. When a relative directory is passed directly (such as `eslint .`):
    1. The relative directory pattern is resolved to a full directory name.
    1. The glob pattern is checked as in step 3.

> **Note:** ESLint will continue to ignore `node_modules` by default.

Reviewer: Just to clarify, would this included any nested `node_modules` (i.e. `**/node_modules`) as well? I would assume it would have to be, since ESLint would look up the directory tree until it finds a configuration file and that could be a parent directory of the project root that contains `node_modules`.

Reviewer: This has been resolved by https://eslint.org/blog/2020/02/whats-coming-in-eslint-7.0.0#updated-default-ignore-patterns

## Alternatives

While there are no alternatives that cover all of the functionality in this RFC, there are alternatives designed to address various parts.

* Both https://github.com/eslint/rfcs/pull/7 and https://github.com/eslint/rfcs/pull/5 specify an alternative method for resolving plugin locations in configs. This attempts to solve the problem with bundling plugins with configs (https://github.com/eslint/eslint/issues/3458). These proposals have the benefit of working within the current configuration system and requiring very few changes from users.
* It is possible to come up with a solution to https://github.com/eslint/eslint/issues/8813 that makes use of `extends`, though this would get a bit complicated if an extended config in an `overrides` section also has `overrides`.
* We could switch the current configuration system over so that all config files are considered to implicitly have `root:true`. This could dramatically simplify configuration searching and merging.

## Open Questions

1. Do we need a command line flag to opt-in to `eslint.config.js` instead of trying to do it alongside the existing configuration system?

> 1. Should `eslint` run without any directories, globs, or filenames, fall back to using the globs in the config to find files to run?

Reviewer: From a user perspective, it could be nice to be able to specify all files being linted directly in the config file and just run `eslint`. 

Would it be a concern that an update to a shared config or plugin could change what files are being linted silently (unless, of course, it results in more errors to be reported)? This is an unlikely scenario (I can't think of one where the intent wouldn't be malicious), but it seems like it could be possible. This is a similar concern to my comment above about allowing `ignores` in plugins and shared configs.

- Start Date: 2019-01-20
- RFC PR: https://github.com/eslint/rfcs/pull/9
- Authors: Nicholas C. Zakas (@nzakas)

> - Contributors: Teddy Katz (@not-an-ardvaark), Toru Nagashima (@mysticatea)

Reviewer 17: ```suggestion
- Contributors: Teddy Katz (@not-an-aardvark), Toru Nagashima (@mysticatea)
```
🙂

Author: Thanks.

```js
const requireIndex = require("requireindex");

module.exports = (context) => {
    const myConfig = {
        files: ["*.js"],
        defs: {
            plugins: {
                custom: {
                    rules: requireIndex("./custom-rules")
                }
            },
        }
        rules: {
            "custom/my-rule": "error"
        }
    };

    if (context.hasRule("some-new-rule")) {
        myConfig.rules["some-new-rule"] = ["error"];
    }

    return myConfig;
};
```

#### Including Function Configs in an Array

A function config can be used anywhere a config object or a config array is valid. That means you can insert a function config as a config array member:

```js
module.exports = [
    (context) => someObject,
    require("eslint-config-myconfig")
];
```

Each function config in an array will be executed with a `context` object when ESLint evaluates the configuration file. This also means that shareable configs can export a function instead of an object or array.

**Note:** If a function config inside of a config array happens to return an array, then those config array items are flattened as with any array-in-array situation.

> ### The `@eslint/config` Utility

Author: `@eslint/utils` is a bit too generic for this purpose (utils for what? How do we know it's just for config?)

  
    translateESLintRC(config)
];
```

#### The `importPlugin()` function

The `importPlugin()` function allows users to automatically load a plugin's rules and processors without separately assigning a namespace. For example:

```js
module.exports = [
    "eslint:recommended",

    // add in eslint-plugin-vue
    importPlugin("vue"),

    // add in eslint-plugin-example
    importPlugin("example")
];
```

This example includes both `eslint-plugin-vue` and `eslint-plugin-example` so that all of the rules are available with the correct namespace and processors are automatically hooked up to the correct `files` pattern.

### Configuration Location Resolution

When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:

1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
1. Otherwise:
    a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
    b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
    c. If a `eslint.config.js` file is found at any point, stop searching and use that file.

>     d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.

Author: We can't guarantee that ESLint will be run under someone's home folder. Most utilities automatically just go up to `/` so it's easier to do that than try to create some different behavior.

* Both https://github.com/eslint/rfcs/pull/7 and https://github.com/eslint/rfcs/pull/5 specify an alternative method for resolving plugin locations in configs. This attempts to solve the problem with bundling plugins with configs (https://github.com/eslint/eslint/issues/3458). These proposals have the benefit of working within the current configuration system and requiring very few changes from users.
* It is possible to come up with a solution to https://github.com/eslint/eslint/issues/8813 that makes use of `extends`, though this would get a bit complicated if an extended config in an `overrides` section also has `overrides`.
* We could switch the current configuration system over so that all config files are considered to implicitly have `root:true`. This could dramatically simplify configuration searching and merging.

## Open Questions

1. Do we need a command line flag to opt-in to `eslint.config.js` instead of trying to do it alongside the existing configuration system?
1. Should `eslint` run without any directories, globs, or filenames, fall back to using the globs in the config to find files to run?

## Frequently Asked Questions

### Can a config be an async function?

> No. Right now it won't be possible to implement a config with an async function because the rest of ESLint is fully synchronous. Once we look at how to make ESLint more asynchronous, we can revisit and allow configs to be created with async functions.

Reviewer 14: Probably can add a note that that #40 is merged so it could eventually be possible when the new format land as stable.

Author: True. I'm still planning on doing this as a later step, but there won't be anything blocking it now.

> ### Why use `module.exports` instead of `module.exports`?

Reviewer 14: the `module.exports` is twice, did you meant `exports.config`?

Author: Oops yes, that's a copy-paste issue from back when the original RFC was included. I've changed since then so this doesn't make any sense.

If any entry in the `files` key is an array, then all of the patterns must match in order for a filename to be considered a match. For example:

```js
module.exports = [
    {
        files: [ ["*.test.*", "*.js"] ],
        rules: {
            semi: ["error", "always"]
        }
    }
];
```

Here, the `files` key specifies two glob patterns. The filename `foo.test.js` would match because it matches both patterns whereas the filename `foo.js` would not match because it only matches one of the glob patterns.

**Note:** This feature is primarily intended for backwards compatibility with eslintrc's ability to specify `extends` in an `overrides` block.

#### Ignoring files

With `eslint.config.js`, there are three ways that files and directories can be ignored:

1. **Defaults** - by default, ESLint will ignore `node_modules` and `.git` directories only. This is different from the current behavior where ESLint ignores `node_modules` and all files/directories beginning with a dot (`.`).
2. **.eslintignore** - the regular ESLint ignore file.
3. **eslint.config.js** - patterns specified in `ignores` keys when `files` is not specified. (See details below.) 

> Anytime `ignores` appears in a config object without `files`, then the `ignores` patterns acts like the `ignorePatterns` key in `.eslintrc` in that the patterns are excluded from all searches before any other matching is done. For example:

Reviewer 12: Low-priority idea: it might be helpful to warn the user if there are any other keys in a config object that has `ignores` but no `files` since the other keys won't do anything.

```bash
eslint
```

This will go into the `eslint.config.js` and use all of the `files` glob patterns to determine which files to lint.

### Replacing `--rulesdir`

In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `plugins` and then specify the rules from a directory. This can be accomplished using the [`requireindex`](https://npmjs.com/package/requireindex) npm package:

```js
const requireIndex = require("requireindex");

module.exports = {
    plugins: {
        custom: {
            rules: requireIndex("./custom-rules")
        }
    },
    rules: {
        "custom/my-rule": "error"
    }
};
```

The `requireIndex()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).

### Function Configs

Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `module.exports` may also be a function that returns an object, such as:

```js

> module.exports = (context) => {

Reviewer 12: I could see at some point wanting to allow this to be an `async` function to allow e.g. runtime `import()`s based on `context`, and this is already contemplated in the FAQ ("Can a config be an async function?"). Even if the initial release requires this to be synchronous, I can't think of anything that would prevent us from adding that in a semver-minor change at a later date since `ESLintConfigArray#normalize()` is already `async`, but I wanted to flag this to see if anyone else thinks of any blockers there.

```js
const { ESLintRCCompat } = require("@eslint/eslintrc");

const eslintrc = new ESLintRCCompat(__dirname);

module.exports = [
    "eslint:recommended",

    // add in eslint-plugin-vue and eslint-plugin-example
    eslintrc.plugins("vue", "example")
];
```

This example includes both `eslint-plugin-vue` and `eslint-plugin-example` so that all of the rules are available with the correct namespace and processors are automatically hooked up to the correct `files` pattern.

#### Applying environments

The `ESLintRCCompat#env()` method allows users to specify an `env` settings as they would in an `.eslintrc`-style config and have globals automatically added. For example:

```js
const { ESLintRCCompat } = require("@eslint/eslintrc");

const eslintrc = new ESLintRCCompat(__dirname);

module.exports = [
    "eslint:recommended",
  
    // load node environment
    eslintrc.env({
        node: true
    })
];
```

### Configuration Location Resolution

When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:

1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
1. Otherwise:

>     a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
>     b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
>     c. If a `eslint.config.js` file is found at any point, stop searching and use that file.
>     d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.

Reviewer 12: Formatting nit: Right now GitHub's Markdown renderer is merging these lines together into one paragraph. If you want to preserve the `<ol>`, this I believe it will translate this into i-iv sub-numbering because of the indentation as it did in the File Pattern Resolution section below:

```suggestion
    1. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
    1. If not found, search up the directory hierarchy looking for `eslint.config.js`.
    1. If a `eslint.config.js` file is found at any point, stop searching and use that file.
    1. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.
```

The following keys are specified the same as in `.eslintrc` files:

* `settings`
* `rules`

The following keys are specified differently than in `.eslintrc` files:

* `plugins` - an object mapping plugin names to implementations. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
* `processor` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
* `languageOptions` - top-level grouping for all options that affect how JavaScript is interpreted by ESLint.
    * `ecmaVersion` - sets the JavaScript version ESLint should use for these files. This value is copied into `parserOptions` if not already present.
    * `sourceType` - sets the source type of the JavaScript code to parse. One of `module`, `script`, or `commonjs`.
    * `parser` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
    * `parserOptions` - an object specifying any additional parameters to be passed to the parser.

>     * `globals` - any additional global variables to add.

Reviewer 10: I have curious, why `processor` is not a language option, differently from `parser`? (or why `parser` is a language option?)

- We may use the `parser` field to lint other than JavaScript. For example, `@typescript-eslint/parser` for TypeScript, `vue-eslint-parser` for Vue.js template files, and maybe for json, for graphql, etc.  
  `babel-eslint` may be a special case; it's a parser for JavaScript.
- We may use the `processor` field to lint other than JavaScript. For example, `eslint-plugin-html`, `eslint-plugin-markdown`, `eslint-plugin-svelte3`, etc.

We are using either field to parse other than JavaScript.

Reviewer 14: Good point. I think it makes sense `processor` to be inside `languageOptions` too for the same reason `parser` is. It processes a language thing. Parser is language thing too.

Author: My thought was that processors don't really do anything that is language-specific. All they do is take text in one format and then split it out into a bunch of smaller pieces of text that can then be parsed. For example, `eslint-plugin-markdown` isn't really an option for JavaScript, it's an option for how to process the file, and then the language settings apply based on the virtual filenames being returned.

Reviewer 10: Parsers are the same. We have used custom parsers to parse other languages than JavaScript. `babel-eslint` is the only exception that it parses JavaScript.

Author: The way I see it, `parser` is an option of the JavaScript language features built into ESLint, whereas `processor` has nothing JavaScript-specific in it. I'd like to leave this as-is, as I can anticipate extending configs in the future to be able to do stuff like this:

```js
[{
    files: ["*.css"],
    language: new CSSImplementation(),
    languageOptions: {
        something: true
    }
},
{
    files: ["*.js"],
    language: new JSImplementation(),
    languageOptions: {
        globals: {}
    }
}]
```

Right now I see our internal stuff as a JavaScript language implementation, and we might ultimately want to swap that out at some point.

Reviewer 10: Hmm.

In the real world, there are multiple parsers that parse other languages than JavaScript and we have used those: `@typescript-eslint/parser`, `vue-eslint-parser`, `@graphql-eslint/parser`, etc. Therefore, because the `parser` is the setting that decides the language, I feel odd if `parser` is a JavaScript option.

Actually, all of `languageOptions` are parser options. The `languageOptions.ecmaVersion` and `languageOptions.sourceType` will be merged into `languageOptions.parserOptions`. The only exception is `languageOptions.globals`, but because the `parser` does scope analysis and returns the scope object, we can think it's a kind of parser options.

That said, I'm OK as-is because you are thinking the mechanism to switch the language that includes the parser and other parts.

        example
    }
};
```

If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown when the configs are normalized because the plugin namespace `example` can only be assigned once.

To work around this problem, the end user can create a separate namespace for the same plugin so that it doesn't conflict with an existing plugin namespace from a shareable config. For example, suppose you want to use `eslint-config-first`, and that has an `example` plugin namespace defined. You'd also like to use `eslint-config-second`, which also has an `example` plugin namespace defined. Trying to use both shareable configs will throw an error because a plugin namespace cannot be defined twice. You can still use both shareable configs by creating a new config from `eslint-config-second` that uses a different namespace. For example:

```js
// get the config you want to extend
const configToExtend = require("eslint-config-second");

// create a new copy (NOTE: probably best to do this with @eslint/config somehow)
const compatConfig = Object.create(configToExtend);
compatConfig.plugins["compat::example"] = require("eslint-plugin-example");

> delete compatConfig.plugins.example;

Reviewer 10: I'm not sure if this is valid because shareable configs may be an array or a function. Is there a more general way?

1. When a glob pattern is passed directly (such as `eslint src/*.js`):
    1. ESLint expands the glob pattern to get a list of files.
    1. Each file is checked individually as in step 1.
1. When a directory is passed directly (such as `eslint src`):
    1. The directory is converted into a glob pattern by appending `**/*` to the directory (such as `src` becomes `src/**/*`).
    1. The glob pattern is checked as in step 2.
1. When a relative directory is passed directly (such as `eslint .`):
    1. The relative directory pattern is resolved to a full directory name.
    1. The glob pattern is checked as in step 3.

**Note:** ESLint will continue to ignore `node_modules` by default.

### Rename `--no-eslintrc` to `--no-config-file` and `useEslintrc` to `useConfigFile`

Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.

### Implementation Details

The implementation of this feature requires the following changes:

> 1. Create a new `ConfigArray` class to manage configs.

Reviewer 10: [ConfigArray](https://github.com/eslint/eslint/blob/master/lib/cli-engine/config-array/config-array.js) class exists already.
Would you create the existing `ConfigArray` instance?

The element type of config arrays is here: [/lib/cli-engine/config-array/config-array.js#L51-L69](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/config-array/config-array.js#L51-L69)

- `name` and `filePath` are just debugging info.
- `criteria` is a tester object to test if given path matches the `files`/`excludedFiles` fields or not. This supports AND operation for `extends` in `overrides`. Currently, the `FileEnumerator` class iterates files where match this tester if a directory path is present.
- `ignorePattern` is a tester object to test if given path matches `.eslintignore` and the `ignorePatterns` field.
- `parser` and `plugins` are loaded objects, with debugging info.
- `processor` is a string, so it needs some tweaks like https://github.com/eslint/eslint/commit/bd5de522e9c2dbeb607f4ac5c7d3d42969c03ac8.
- `env`, `globals`, `noInlineConfig`, `parserOptions`, `reportUnusedDisableDirectives`, `root`, and `settings` are the configured values as-is.
- [`extractConfig(filePath)`](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/config-array/config-array.js#L470-L485) method merges the array elements to a final configuration.

It will reduce the amount of code that we need to modify because we don't need to modify the use side of the config arrays mostly.

----

Currently, the creation logic of the config arrays are in [ConfigArrayFactory](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/config-array-factory.js) and [CascadingConfigArrayFactory](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/cascading-config-array-factory.js). The `ConfigArrayFactory` loads a file then creates a config array. The `CascadingConfigArrayFactory` concatenates (and caches) config arrays of all of CLI options, found config files, and `.eslintignore`.

The implementation will be alternate to those factories.

Author: Yes, the creation logic will be completely different from what we have today. I've renamed the class to `ESLintConfigArray` to distinguish from the current implementation (which didn't exist when I created this RFC).

Reviewer 10: My intention was, creation logic is two existing factories and the new config system can use intermediate representation `CofnigArray` class as-is.

Author: Oh I see. It's unfortunate that you didn't let me know that ahead of time. I don't believe we can mix the two implementations at this point, nor do I think it would be a good idea.

Reviewer 10: > I don't believe we can mix the two implementations at this point

I can implement it. Because the intermediate object has all of the required features to represent the new config and it's completely separated from the creation logic of the current config.

> nor do I think it would be a good idea

If the two systems have the same intermediate representation, we don't need to modify the use side of the intermediate representation. The use side doesn't need to care where the config array came from. It will simplify our codebase.

1. Create a `--no-config-file` CLI option and alias it to `--no-eslintrc` for backwards compatibility.
1. Create a `useConfigFile` option for `CLIEngine`. Alias `useEslintrc` to this option for backwards compatibility.
1. In `CLIEngine#executeOnFiles()`:
    1. Check for existence of `eslint.config.js`, and if found, opt-in to new behavior.
    1. Create a `ConfigArray` to hold the configuration information and to determine which files to lint (in conjunction with already-existing `globUtils`)
    1. Rename the private functions `processText()` and `processFiles()` to `legacyProcessText()` and `legacyProcessFiles()`; create new versions with the new functionality named `processText()` and `processFiles()`. Use the appropriate functions based on whether or not the user has opted-in.

>     1. Update `Linter#verify()` to check for objects on keys that now support objects instead of strings (like `parser`).

Reviewer 10: I think we need a new option for `Linter#verify()` in order to disable the processing of `env`. All of the `env` processing is in `Linter#verify()`.

Author: Good catch! I'll get that incorporated.

1. Searching for `eslint.config.js` starts from the current working directory and continues up
1. All `eslint.config.js` files are treated as if they have `root: true`
1. There is no automatic merging of config files
1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
1. An `@eslint/eslintrc` utility is provided to aid with backwards compatibility
1. Remove the concept of environments (`env`) (including `--env`, `/*eslint-env*/`, and `env` in config files)
1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
1. Remove `--rulesdir`
1. Remove `--ext`
1. Remove `--resolve-plugins-relative-to`
1. Replace `context.parserPath` with `context.parser` (the path to the parser is no longer valid)
1. Allow plugins to specify a `parsers` key so parsers can be referenced the same as rules and processors in plugins.

### The `eslint.config.js` File

The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports an object: 

```js
module.exports = {
    name: "name",
    files: ["*.js"],
    ignores: ["*.test.js"],

>     settings: {},

Reviewer: As I've been thinking about this more, the property name `settings` feels out of place. Is there any way we can make give it a less vague name? Maybe `pluginOptions`?

Author: I'm not sure if changing the name has much of a benefit vs. the confusion it could potentially cause, but I don't feel very strongly about it.

Author: So after looking into this some more, I think changing `settings` to something else will make backwards compatibility more challenging (these are exposed through `context.settings` in rules). While I agree the name isn't ideal, I think it would cause more trouble to change it than just to let it be.

Reviewer: That's fair :+1:

## Motivation

The ESLint configuration file format (`.eslintrc`) hasn't changed much since ESLint was first released. As we have continued to add new features, the configuration file format has continued to be extended and evolve into a state where it is the most complicated part of ESLint. The `.eslintrc` way of doing things means a variety of different merging strategies (cascading, extending, overriding), loading strategies (for plugins, processors, parsers), and naming conventions (`eslint-config-*`, `eslint-plugin-*`).

Although `.eslintrc` has proven to be a very robust format, it is also limited and unable to easily accommodate feature requests that the community favors, such as:

1. [bundling plugins with configs](https://github.com/eslint/eslint/issues/3458)
1. [specifying ignore patterns in the config](https://github.com/eslint/eslint/issues/10891)
1. [specifying `--ext` information in the config](https://github.com/eslint/eslint/issues/11223)

> 1. [using `extends` in `overrides`](https://github.com/eslint/eslint/issues/8813) 

Author: I'd like to keep these here to reflect why the RFC was written in the first place.

> 1. [using `extends` in `overrides`](https://github.com/eslint/eslint/issues/8813) 

Reviewer 10: Isn't "unable to easily accommodate feature requests that the community favors" the primary motivation of this RFC? This RFC has huge migration pain as every in the community has to migrate their configuration, so the balanced motivation is very important. Why does the motivation list the supported features as unable to support?

> 1. [using `extends` in `overrides`](https://github.com/eslint/eslint/issues/8813) 

Author: > Why does the motivation list the supported features as unable to support?

It doesn't. As you stated, it says "unable to easily accommodate." I would say that's a correct statement given how long those requests were open before they actually got implemented. If it had been easy (keeping in mind that we got requests prior to the mentioned issues, as well).

And as I already said, I think it's helpful for the RFC's original motivation to remain so we have context in the future as to why it was written in the first place. At the time it was written, none of the features were implemented and no one had even attempted them. I think that's important to remember.

As a compromise, I'll add a note stating these three were implemented in `.eslintrc` after this RFC was written.

1. Remove `--rulesdir`
1. Remove `--ext`
1. Remove `--resolve-plugins-relative-to`
1. Replace `context.parserPath` with `context.parser` (the path to the parser is no longer valid)
1. Allow plugins to specify a `parsers` key so parsers can be referenced the same as rules and processors in plugins.

### The `eslint.config.js` File

The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports an object: 

```js
module.exports = {
    name: "name",
    files: ["*.js"],
    ignores: ["*.test.js"],
    settings: {},
    languageOptions: {
        ecmaVersion: 2020,
        sourceType: "module",
        globals: {},
        parser: object || "string",
        parserOptions: {},
        linterOptions: {
            reportUnusedDisableDirectives: "string"
        }
    }
    processor: object || "string",
    plugins: {}
    rules: {}
};
```

The following keys are new to the `eslint.config.js` format:

* `name` - Specifies the name of the config object. This is helpful for printing out debugging information and, while not required, is recommended for that reason.
* `files` - Determines the glob file patterns that this configuration applies to.

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Author: `ignores` acts like `excludedFiles` when `files` is present.

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Reviewer 10: Thanks. So does the `ignores` property have different meanings between the `files` property is present and not?

- If the `files` is present, ESLint uses the `ignores` as `excludedFiles` before it merges config array elements.
- Otherwise, ESLint merges the `ignores` as well, then it doesn't lint the files that match the merged `ignores` at all.

If so, I don't think it's good behavior.

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Author: It's really the same meaning:

* if `files` is present: match patterns in `files` and then exclude things in `ignores`.
* if `files` is not present: assume `files` is "*" and then exclude things in `ignores`.

So the matching works the same way in both cases, we just assume you meant for `files` to apply to everything if it's not present.

As a reminder: the prototype already took this into account and things worked fine.

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Reviewer 10: No. I quote this line.

> Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

1. The "Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file." part describes the functionality of `.eslintignore` and the `ignorePatterns` in eslintrc.
1. The "The files specified by this array of glob patterns are subtracted from the files specified in `files`." part describes the functionality of the `excludedFiles` field of eslintrc.
1. The "If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files." part describes the functionality of `.eslintignore` and the `ignorePatterns` in eslintrc again.

I feel this is really confusing.

The `excludedFiles` and `ignorePatterns` (`.eslintignore`) fields in eslintrc are pretty different.

```js
module.exports = {
    {
        files: ["*.js"],
        rules: { /* rules for all code */ }
    },
    {
        // files: ["*.js"],
        ignores: ["*.spec.js"],
        rules: { /* rules for production js code */ }
    },
}
```

With the above config, the second entry doesn't have the `files` field, so I assume it works as same as `.eslintignore`. Therefore, ESLint doesn't lint `*.spec.js` at all. Because the purpose of `.eslintignore` is to specify ignored files/dirs such as `node_modules`, `dist`, or `build_cache`.

On the other hand, if I assume the `ignores` field behaves as same as the `excludedFiles` field of eslintrc, with the above config, ESLint lints `*.spec.js` files with using only the first entry. ESLint doesn't ignore `*.spec.js` files.

----

> - if files is present: match patterns in files and then exclude things in ignores.
> - if files is not present: assume files is "*" and then exclude things in ignores.

It's the description of `excludedFiles`. It's pretty different from `.eslintignore`/`ignorePatterns`.

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Reviewer 10: > As a reminder: the prototype already took this into account and things worked fine.

The prototype doesn't work for me.

```
> npm i eslint/eslint#simpleconfig
> eslint test.js
Error: Key "ignores" requires keys "files".
    at ObjectSchema.validate (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\@humanwhocodes\object-schema\src\object-schema.js:157:27)
    at C:\Users\t-nagashima.AD\dev\sandbox\node_modules\@humanwhocodes\object-schema\src\object-schema.js:116:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\@humanwhocodes\object-schema\src\object-schema.js:114:24)
    at C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\simpleconfig\config-array.js:114:27
    at Array.reduce (<anonymous>)
    at ConfigArray.getConfig (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\simpleconfig\config-array.js:113:32)
    at processText (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\cli-engine.js:183:33)
    at processFile (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\cli-engine.js:246:12)
    at C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\cli-engine.js:634:40
```

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Author: > On the other hand, if I assume the ignores field behaves as same as the excludedFiles field of eslintrc, with the above config, ESLint lints *.spec.js files with using only the first entry. ESLint doesn't ignore *.spec.js files.

Just to make sure I understand: you're saying that `excludedFiles` does not require the use of `files` for overrides? In that case, the override will always apply *except* for the files in `excludedFiles`?

> * `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

Reviewer 10: > Just to make sure I understand: you're saying that excludedFiles does not require the use of files for overrides? In that case, the override will always apply except for the files in excludedFiles?

No. We cannot omit the `files` field in `overrides` of the current config system. I meant that, if I interpret that the `ignores` corresponds to the current `excludedFiles`, ...

In the current config system, `ignorePatterns` and `excludedFiles` are different stuff:

- `ignorePatterns`:
  - A list of gitignore-style patterns. (therefore, we can read .gitignore and set it.)
  - All of the `ignorePatterns` field are merged. It's used *after* all config array elements are merged to one.
  - It supports negative patterns to unignore other ignore patterns. For example, `!foo.js` will unignore `foo.js` even if a shareable config is configured to ignore it.
- `excludedFiles`:
  - A list of glob patterns.
  - Every `excludedFiles` field is used to check the adoption of each `overrides` entry. It's used *before* all config array elements are merged to one.
  - It doesn't support negative patterns.

This RFC says a mix of those for the `ignores` field. Therefore, I'm confusing.

- If `files` field exists, the `ignores` field is used for the adoption of the config array element and isn't merged for the check of ignored files. It's glob patterns as the RFC says.
- If `files` field doesn't exist, the `ignores` field doesn't affect the adoption of the config array element and is merged for the check ignored files. It's a list of gitignore-style patterns and supports negative patterns, as same as `.eslintignore`/`ignorePatterns`.

* `processor` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
* `languageOptions` - top-level grouping for all options that affect how JavaScript is interpreted by ESLint.
    * `ecmaVersion` - sets the JavaScript version ESLint should use for these files. This value is copied into `parserOptions` if not already present.
    * `sourceType` - sets the source type of the JavaScript code to parse. One of `module`, `script`, or `commonjs`.
    * `parser` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
    * `parserOptions` - an object specifying any additional parameters to be passed to the parser.
    * `linterOptions` - an object for linter-specific settings
        * `reportUnusedDisableDirectives` - new location for the same option name.
    * `globals` - any additional global variables to add.

Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `eslint.config.js`, these are all objects or strings (referencing an object in a plugin), requiring users to manually specify the objects to use.

The following keys are invalid in `eslint.config.js`:

> * `extends` - replaced by config arrays

Author: > Probably, we should not use spread properties to extend shareable configs because the shareable configs have different forms: object, array, function. Therefore, extends or something like is still needed.

Yeah, that's a good point. It probably makes sense to add back in `extends` for this purpose.

> * `extends` - replaced by config arrays

Author: I took a look at specifying this, and I'm not sure `extends` will work. As an example, suppose we have a shareable config called `eslint-config-example` that is defined as this:

```js
module.exports = [
    {
        files: ["*.js"],
        languageOptions: { sourceType: "module" },
        rules: {
            semi: ["error", "always"]
        }
    },
    {
        files: ["*.cjs"],
        languageOptions: { sourceType: "commonjs" },
        rules: {
            semi: ["error", "never"]
        }
    }
];
```

If in my config I do this:

```js
module.exports = [
    {
        files: ["*.js"],
        extends: [ require("eslint-config-example") ]
    }
];
```

If we `&&` the `files` fields, then we'll end up with `sourceType` as `"commonjs"` (because it is last), which is incorrect. I'm not sure the `eslintrc` algorithm for `overrides` makes sense here.

After looking into this some more, I'm thinking we should leave off `extends` for now and see how people end up using this in the wild. We can always add something later to handle this case if it becomes common enough.

In the meantime, we can always recommend that shareable configs should not include `files` or `ignores` if they are meant to be used outside of the current project.

> * `extends` - replaced by config arrays

Reviewer 10: > If we && the files fields, then we'll end up with sourceType as "commonjs" (because it is last)

Why? a `*.js` file doesn't match `"*.cjs"`.

> * `extends` - replaced by config arrays

Author: Oh okay, I got a bit confused by your description. Now I understand, it would only match if the filename matched both `*.js` and `*.cjs` at the same time.

I need to think about this some more. Trying to hack an AND operator into `files` seems a bit messy to me.

It does seem like we need some way of extending `files` patterns from shareable configs, I'm just not sure what we currently do in `overrides` is the right model for this new format.

> * `extends` - replaced by config arrays

Author: After thinking about this for a bit longer, it seems there are actually two different use cases we need to support:

1. The current `overrides` behavior, which implements an AND between grep patterns. We will need to support this in order to easily allow backwards compatibility with how eslintrc currently works.
2. The case that @kaicataldo was talking about can't actually be solved with an AND, it must be an OR. That is, if a config already has a `files`, then we want to also use that config for a different pattern.

In terms of priority, I think that the AND case is the one that needs solving immediately (I do have an idea for that) for backwards compatibility cases. I think we can defer the OR case for now by recommending that shareable configs designed to be used this way should be a single object without a `files` property, and that would allow us to do patterns like use a spread. Also, because the config file is JavaScript, people will be free to manipulate objects in whatever ways they want, so I think it's less critical for us to provide an out-of-the-box solution from the start.

> * `extends` - replaced by config arrays

Author: Okay, I've added in a bit about how to accomplish AND in the `files` array.

I think we can successfully solve the OR case with documentation to start, and then we can figure out the patterns people are using most frequently and see if we need to add anything more.

```js
const requireIndex = require("requireindex");

module.exports = {
    plugins: {
        custom: {
            rules: requireIndex("./custom-rules")
        }
    },
    rules: {
        "custom/my-rule": "error"
    }
};
```

The `requireIndex()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).

### Function Configs

Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `module.exports` may also be a function that returns an object, such as:

```js
module.exports = (context) => {

    // do something

    return {
        files: ["*.js"],
        rules: {
            "semi": ["error", "always"]
        }
    };

};
```

The `context` object has the following members:

> * `name` - the name of the application being used

Reviewer 14: Not clear to me what that "name of the application" is. ESLint? Or the config name?

Author: For ESLint it would be "eslint". If, for example, Standard decided to support some configuration, they could change it to "standard".

* `version` - the version of ESLint being used

> * `hasRule(ruleId)` - determine if the given rule is in the core

Reviewer 14: I don't see how this could be done, since to know the config (we need it to check ruleId against it) and so pass the context we need the config (returned from that function).

The only thing i can think of is passing the (fully resolved) config to that `hasRule(ruleId, config)`.

Author: `hasRule()` is intended for core rules only, not for plugin rules. The core rules are known ahead of time.

Author: That said, maybe `hasRule()` is confusing enough that we should leave it out. It's certainly not necessary for this RFC.

Author: I've gone and removed that method. Thanks for the feedback.

**Note:** If a function config inside of a config array happens to return an array, then those config array items are flattened as with any array-in-array situation.

### The `@eslint/eslintrc` Utility

To allow for backwards compatibility with existing configs and plugins, an `@eslint/eslintrc` utility is provided. The package exports the following classes:

* `ESLintRCCompat` - a class to help convert `.eslintrc`-style configs into the correct format.

```js
class ESLintRCCompat {
    
    constructor(baseDir) {}
    
    config(configObjct) {}
    plugins(...pluginNames) {}
    extends(...sharedConfigNames) {}
    env(envObject) {}

}
```

#### Importing existing configs

The `ESLintRCCompat#extends()` function allows users to specify an existing `.eslintrc` config location in the same format that used in the `.eslintrc` `extends` key. Users can pass in a filename, a shareable config name, or a plugin config name and have it converted automatically into the correct format. For example:

```js
const { ESLintRCCompat } = require("@eslint/eslintrc");

const eslintrc = new ESLintRCCompat(__dirname);

module.exports = [

>     "eslint:recommended",

Reviewer 14: Does that translate internally to something like the following

```js
module.exports = [
  {
    plugins: {
      'eslint:recommended': require('eslint/path/to/internal/rules')
    }
  }
]
```

Author: No, plugin declarations create namespaces, not definitions. The string value would just need to be replaced with an object during normalization.

Reviewer 14: That's what I'm talking about. To what will be normalized the `'eslint:recommended'` string?

Here, it is the `ruledefs` that assigns the name `react` to the rules from `eslint-plugin-react`. The reference to `react/` in a rule will always look up that value in the `ruledefs` key.

> **Note:** If a config is merged with another config that already has a `ruledefs` namespace defined, then an error is thrown. In this case, if a config already has a `react` namespace, then attempting to combine with another config that has a `react` namespace will throw an error. This is to ensure the meaning of `namespace/rule` remains consistent.

Reviewer 9: Should this say "the same" (or "a namespace with the same name as the config to be merged", or similar)? The way I'm currently reading the sentence, it could be interpreted as *any* two configs that have *anything* in ruledefs will error, but I think it's supposed to be if they export ruledefs under the same namespace name.

Author: Yes, that would be a much clearer explanation of my intent here. Thanks, I'll update that.

When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.

#### Using AND patterns for files

If any entry in the `files` key is an array, then all of the patterns must match in order for a filename to be considered a match. For example:

```js
module.exports = [
    {
        files: [ ["*.test.*", "*.js"] ],
        rules: {
            semi: ["error", "always"]
        }

>     }

Reviewer 10: How can I apply a shareable config to only specific files?

Author: As mentioned previously, for the time being, we need to do this by documenting that shareable configs that want to be applied to multiple file patterns should be published without a `files` field specified. That way, consumers can just do something like this:

```js
const sharedConfig = require("eslint-config-shared");

module.exports = [
    {
        files: ["*.js"],
        ....sharedConfig
    },
    {
        files: ["*.mjs"],
        ...sharedConfig
    }
];
```

Because shared configs don't need to follow a specific pattern, there are any number of ways developers can create shared configs to allow end users to easily mix-and-match. If we see common patterns emerging, we can always revisit and provide some utilities to help.

Reviewer 10: By this RFC, config can be an array, a function, or an object. We cannot use spread property syntax to merge those.

Author: Again, we can solve this with documentation in the short-term. No shareable config is an array or a function right now, so we can tell people that if they want to make their config usable for any file pattern, make sure it's either an object or keep it in `eslintrc` format.

Once we get some feedback and see how people are using the new config system, we can figure out what other utilities we should add to make people's lives easier.

Reviewer 10: > No shareable config is an array or a function right now

That's not practical because most shareable configs are an array to use `extends`. It's pretty common use-case that shareable configs extend other shareable configs and plugin's recommended config.

Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.

### Implementation Details

The implementation of this feature requires the following changes:

1. Create a new `ConfigArray` class to manage configs.
1. Create a `--no-config-file` CLI option and alias it to `--no-eslintrc` for backwards compatibility.

> 1. Create a `useConfigFile` option for `CLIEngine`. Alias `useESLintRC` to this option for backwards compatibility.

Reviewer 9: TODO: Suggestion for `useEslintrc`

1. In `CLIEngine#executeOnFiles()`:
    1. Check for existence of `eslint.config.js`, and if found, opt-in to new behavior.

>     1. Create a `ConfigArray` to hold the configuration information and to determine which files to lint (in conjuction with already-existing `globUtils`)

Reviewer 9: TODO: Typo on "conjuction"

**Note:** If a function config inside of a config array happens to return an array, then those config array items are flattened as with any array-in-array situation.

> ### The `@eslint/eslintrc` Utility

Reviewer 14: :tada:
